diff --git a/tasks/zeroshot_gpt/evaluate.py b/tasks/aquila/evaluate.py
similarity index 89%
rename from tasks/zeroshot_gpt/evaluate.py
rename to tasks/aquila/evaluate.py
index e42c776e8..b63592ac3 100644
--- a/tasks/zeroshot_gpt/evaluate.py
+++ b/tasks/aquila/evaluate.py
@@ -6,16 +6,16 @@ import math
 
 import torch
 
-from megatron.training import get_args
-from megatron.training import print_rank_0, is_last_rank
-from megatron.training import get_tokenizer
+from megatron import get_args
+from megatron import print_rank_0, is_last_rank
+from megatron import get_tokenizer
 from megatron.core import parallel_state, tensor_parallel
-from megatron.training.checkpointing import load_checkpoint
-from megatron.legacy.model import GPTModel
+from megatron.checkpointing import load_checkpoint
+from megatron.model import GPTModel
 from megatron.training import get_model
-from megatron.training.utils import get_ltor_masks_and_position_ids, unwrap_model
+from megatron.utils import get_ltor_masks_and_position_ids, unwrap_model
 from megatron.core.pipeline_parallel.p2p_communication import recv_forward, send_forward
-from megatron.training.arguments import core_transformer_config_from_args
+from megatron.arguments import core_transformer_config_from_args
 from tasks.finetune_utils import build_data_loader
 
 from .datasets import build_dataset
@@ -96,7 +96,9 @@ def forward_step(batch, model, eval_metric, config):
                 output.contiguous().float(), labels.contiguous())
             loss = torch.sum(
                 losses.view(-1) * loss_mask.contiguous().view(-1).float())
-            return loss
+            loss_mask = torch.sum(
+                loss_mask.contiguous().view(-1).float())
+            return loss / loss_mask
 
         # For accuracy, return the number of correctly predicted samples.
         if eval_metric == 'accuracy':
@@ -180,13 +182,11 @@ def main():
     args = get_args()
 
     if args.num_layers_per_virtual_pipeline_stage is not None:
-        print("Interleaved pipeline schedule is not yet supported for text generation.")
+        print("Interleaved pipeline schedule is not yet supported for evaluation.")
         exit()
 
-    if args.task == 'LAMBADA':
-        eval_metric = 'accuracy'
-    elif args.task == 'WIKITEXT103':
-        eval_metric = 'loss'
+    if args.eval_metric is not None:
+        eval_metric = args.eval_metric
     else:
         raise NotImplementedError('{} task is not implemented.'.format(
             args.task))

