diff --git a/vllm/entrypoints/openai/tool_parsers/internlm2_tool_parser.py b/vllm/entrypoints/openai/tool_parsers/internlm2_tool_parser.py
index 37c360145..e5dcdf9a0 100644
--- a/vllm/entrypoints/openai/tool_parsers/internlm2_tool_parser.py
+++ b/vllm/entrypoints/openai/tool_parsers/internlm2_tool_parser.py
@@ -8,7 +8,7 @@ from typing import Union
 import partial_json_parser
 from partial_json_parser.core.options import Allow
 
-from vllm.entrypoints.chat_utils import make_tool_call_id
+from vllm.entrypoints.chat_utils import random_tool_call_id
 from vllm.entrypoints.openai.protocol import (ChatCompletionRequest,
                                               DeltaFunctionCall, DeltaMessage,
                                               DeltaToolCall,
@@ -35,12 +35,12 @@ class Internlm2ToolParser(ToolParser):
             self, request: ChatCompletionRequest) -> ChatCompletionRequest:
         if request.tools and request.tool_choice != 'none':
             # do not skip special tokens because internlm use the special
-            # tokens to indicate the start and end of the tool calls
+            # tokens to indicated the start and end of the tool calls
             # information.
             request.skip_special_tokens = False
         return request
 
-    def get_arguments(self, obj):
+    def get_argments(self, obj):
         if "parameters" in obj:
             return obj.get("parameters")
         elif "arguments" in obj:
@@ -60,8 +60,8 @@ class Internlm2ToolParser(ToolParser):
         if '<|action_start|>' not in current_text:
             self.position = len(current_text)
             return DeltaMessage(content=delta_text)
-        # if the tool call is sent, return an empty delta message
-        # to make sure the finish_reason will be sent correctly.
+        # if the tool call is sended, return a empty delta message
+        # to make sure the finish_reason will be send correctly.
         if self.current_tool_id > 0:
             return DeltaMessage(content='')
 
@@ -89,7 +89,7 @@ class Internlm2ToolParser(ToolParser):
         try:
             parsable_arr = action
 
-            # tool calls are generated in an object in internlm2
+            # tool calls are generated in an object in inernlm2
             # it's not support parallel tool calls
             try:
                 tool_call_arr: dict = partial_json_parser.loads(
@@ -107,7 +107,7 @@ class Internlm2ToolParser(ToolParser):
                     delta = DeltaMessage(tool_calls=[
                         DeltaToolCall(index=self.current_tool_id,
                                       type="function",
-                                      id=make_tool_call_id(),
+                                      id=random_tool_call_id(),
                                       function=DeltaFunctionCall(
                                           name=function_name).model_dump(
                                               exclude_none=True))
@@ -119,9 +119,9 @@ class Internlm2ToolParser(ToolParser):
             # now we know we're on the same tool call and we're streaming
             # arguments
             else:
-                prev_arguments = self.get_arguments(
+                prev_arguments = self.get_argments(
                     self.prev_tool_call_arr[self.current_tool_id])
-                cur_arguments = self.get_arguments(tool_call_arr)
+                cur_arguments = self.get_argments(tool_call_arr)
 
                 # not arguments generated
                 if not cur_arguments and not prev_arguments:
@@ -170,7 +170,7 @@ class Internlm2ToolParser(ToolParser):
             # check to see if the name is defined and has been sent. if so,
             # stream the name - otherwise keep waiting
             # finish by setting old and returning None as base case
-            tool_call_arr["arguments"] = self.get_arguments(tool_call_arr)
+            tool_call_arr["arguments"] = self.get_argments(tool_call_arr)
             self.prev_tool_call_arr = [tool_call_arr]
             return delta
         except Exception:

