diff --git a/vllm/entrypoints/cli/run_batch.py b/vllm/entrypoints/cli/run_batch.py
index e669464bf..6bdd3b63c 100644
--- a/vllm/entrypoints/cli/run_batch.py
+++ b/vllm/entrypoints/cli/run_batch.py
@@ -1,41 +1,37 @@
 # SPDX-License-Identifier: Apache-2.0
 # SPDX-FileCopyrightText: Copyright contributors to the vLLM project
 
-from __future__ import annotations
-
 import argparse
 import asyncio
-import importlib.metadata
-import typing
-
-from vllm.entrypoints.cli.types import CLISubcommand
-from vllm.entrypoints.utils import VLLM_SUBCMD_PARSER_EPILOG
-from vllm.logger import init_logger
 
-if typing.TYPE_CHECKING:
-    from vllm.utils import FlexibleArgumentParser
+from prometheus_client import start_http_server
 
-logger = init_logger(__name__)
+from vllm.entrypoints.cli.types import CLISubcommand
+from vllm.entrypoints.logger import logger
+from vllm.entrypoints.openai.run_batch import main as run_batch_main
+from vllm.entrypoints.openai.run_batch import make_arg_parser
+from vllm.entrypoints.utils import (VLLM_SUBCMD_PARSER_EPILOG,
+                                    show_filtered_argument_or_group_from_help)
+from vllm.utils import FlexibleArgumentParser
+from vllm.version import __version__ as VLLM_VERSION
 
 
 class RunBatchSubcommand(CLISubcommand):
     """The `run-batch` subcommand for vLLM CLI."""
-    name = "run-batch"
+
+    def __init__(self):
+        self.name = "run-batch"
+        super().__init__()
 
     @staticmethod
     def cmd(args: argparse.Namespace) -> None:
-        from vllm.entrypoints.openai.run_batch import main as run_batch_main
-
-        logger.info("vLLM batch processing API version %s",
-                    importlib.metadata.version("vllm"))
+        logger.info("vLLM batch processing API version %s", VLLM_VERSION)
         logger.info("args: %s", args)
 
         # Start the Prometheus metrics server.
         # LLMEngine uses the Prometheus client
         # to publish metrics at the /metrics endpoint.
         if args.enable_metrics:
-            from prometheus_client import start_http_server
-
             logger.info("Prometheus metrics enabled")
             start_http_server(port=args.port, addr=args.url)
         else:
@@ -46,10 +42,8 @@ class RunBatchSubcommand(CLISubcommand):
     def subparser_init(
             self,
             subparsers: argparse._SubParsersAction) -> FlexibleArgumentParser:
-        from vllm.entrypoints.openai.run_batch import make_arg_parser
-
         run_batch_parser = subparsers.add_parser(
-            self.name,
+            "run-batch",
             help="Run batch prompts and write results to file.",
             description=(
                 "Run batch prompts using vLLM's OpenAI-compatible API.\n"
@@ -58,8 +52,9 @@ class RunBatchSubcommand(CLISubcommand):
             "vllm run-batch -i INPUT.jsonl -o OUTPUT.jsonl --model <model>",
         )
         run_batch_parser = make_arg_parser(run_batch_parser)
-        run_batch_parser.epilog = VLLM_SUBCMD_PARSER_EPILOG.format(
-            subcmd=self.name)
+        show_filtered_argument_or_group_from_help(run_batch_parser,
+                                                  "run-batch")
+        run_batch_parser.epilog = VLLM_SUBCMD_PARSER_EPILOG
         return run_batch_parser
 
 

