diff --git a/tools/report_build_time_ninja.py b/tools/report_build_time_ninja.py
index 7386cdd9f..7368ae953 100644
--- a/tools/report_build_time_ninja.py
+++ b/tools/report_build_time_ninja.py
@@ -116,7 +116,7 @@ def ReadTargets(log, show_all):
             # If ninja.exe is rudely halted then the .ninja_log file may be
             # corrupt. Silently continue.
             continue
-        start, end, _, name, cmdhash = parts  # Ignore restart.
+        start, end, _, name, cmdhash = parts  # Ignore restat.
         # Convert from integral milliseconds to float seconds.
         start = int(start) / 1000.0
         end = int(end) / 1000.0

