diff --git a/vllm/model_executor/layers/quantization/gptq_bitblas.py b/vllm/model_executor/layers/quantization/gptq_bitblas.py
index 646229258..78e0f59fa 100644
--- a/vllm/model_executor/layers/quantization/gptq_bitblas.py
+++ b/vllm/model_executor/layers/quantization/gptq_bitblas.py
@@ -3,7 +3,6 @@
 from typing import Any, Optional
 
 import torch
-from packaging import version
 from torch.nn.parameter import Parameter
 
 from vllm.logger import init_logger
@@ -64,8 +63,7 @@ class GPTQBitBLASConfig(QuantizationConfig):
 
         try:
             import bitblas
-            if version.parse(bitblas.__version__) < version.parse(
-                    MINIMUM_BITBLAS_VERSION):
+            if bitblas.__version__ < MINIMUM_BITBLAS_VERSION:
                 raise ImportError(
                     "bitblas version is wrong. Please "
                     f"install bitblas>={MINIMUM_BITBLAS_VERSION}")
@@ -83,7 +81,6 @@ class GPTQBitBLASConfig(QuantizationConfig):
             # (since we have only one group per output channel)
             desc_act = False
 
-        super().__init__()
         self.weight_bits = weight_bits
         self.group_size = group_size
         self.desc_act = desc_act
@@ -265,9 +262,9 @@ class GPTQBitBLASLinearMethod(LinearMethodBase):
             scales ('scales'), and zeros ('zeros').
 
         Raises:
-            ValueError: If `params_dtype` is not `torch.float16` or if the input
-                size per partition is not divisible by the group size
-                in `quant_config`.
+            ValueError: If `params_dtype` is not `torch.float16` or 
+            if the input size per partition is not divisible by the 
+            group size in `quant_config`.
         """
         if params_dtype != torch.float16:
             raise ValueError("Parameter data type must be torch.float16, "

