diff --git a/vllm/attention/__init__.py b/vllm/attention/__init__.py
index 1b37bd1f6..344040586 100644
--- a/vllm/attention/__init__.py
+++ b/vllm/attention/__init__.py
@@ -2,7 +2,9 @@
 # SPDX-FileCopyrightText: Copyright contributors to the vLLM project
 
 from vllm.attention.backends.abstract import (AttentionBackend,
-                                              AttentionMetadata, AttentionType)
+                                              AttentionMetadata,
+                                              AttentionMetadataBuilder,
+                                              AttentionState, AttentionType)
 from vllm.attention.layer import Attention
 from vllm.attention.selector import get_attn_backend
 
@@ -11,5 +13,8 @@ __all__ = [
     "AttentionBackend",
     "AttentionMetadata",
     "AttentionType",
+    "AttentionMetadataBuilder",
+    "Attention",
+    "AttentionState",
     "get_attn_backend",
 ]

