diff --git a/examples/post_training/modelopt/generate.sh b/examples/post_training/modelopt/generate.sh
index 16f6c1187..58578bb3d 100644
--- a/examples/post_training/modelopt/generate.sh
+++ b/examples/post_training/modelopt/generate.sh
@@ -23,6 +23,7 @@ if [ -z ${PROMPTS_PATH} ]; then
     ${LAUNCH_SCRIPT} ${SCRIPT_DIR}/generate.py \
         ${MODEL_ARGS} \
         --tensor-model-parallel-size ${TP} \
+        --expert-tensor-parallel-size ${ETP} \
         --expert-model-parallel-size ${EP} \
         --pipeline-model-parallel-size ${PP} \
         --tokenizer-model ${TOKENIZER_MODEL} \
@@ -34,6 +35,7 @@ else
     ${LAUNCH_SCRIPT} ${SCRIPT_DIR}/generate.py \
         ${MODEL_ARGS} \
         --tensor-model-parallel-size ${TP} \
+        --expert-tensor-parallel-size ${ETP} \
         --expert-model-parallel-size ${EP} \
         --pipeline-model-parallel-size ${PP} \
         --tokenizer-model ${TOKENIZER_MODEL} \

