diff --git a/vllm/model_executor/models/__init__.py b/vllm/model_executor/models/__init__.py
index 4ccba64f2..27c169d2d 100644
--- a/vllm/model_executor/models/__init__.py
+++ b/vllm/model_executor/models/__init__.py
@@ -1,11 +1,10 @@
 # SPDX-License-Identifier: Apache-2.0
 # SPDX-FileCopyrightText: Copyright contributors to the vLLM project
 
-from .interfaces import (HasInnerState, SupportsLoRA, SupportsMRoPE,
-                         SupportsMultiModal, SupportsPP, SupportsTranscription,
-                         SupportsV0Only, has_inner_state, supports_lora,
-                         supports_mrope, supports_multimodal, supports_pp,
-                         supports_transcription, supports_v0_only)
+from .interfaces import (HasInnerState, SupportsLoRA, SupportsMultiModal,
+                         SupportsPP, SupportsV0Only, has_inner_state,
+                         supports_lora, supports_multimodal, supports_pp,
+                         supports_v0_only)
 from .interfaces_base import (VllmModelForPooling, VllmModelForTextGeneration,
                               is_pooling_model, is_text_generation_model)
 from .registry import ModelRegistry
@@ -22,12 +21,8 @@ __all__ = [
     "supports_lora",
     "SupportsMultiModal",
     "supports_multimodal",
-    "SupportsMRoPE",
-    "supports_mrope",
     "SupportsPP",
     "supports_pp",
-    "SupportsTranscription",
-    "supports_transcription",
     "SupportsV0Only",
     "supports_v0_only",
 ]

