diff --git a/vllm/model_executor/models/interfaces_base.py b/vllm/model_executor/models/interfaces_base.py
index 8fdf70e35..4a1ea74a2 100644
--- a/vllm/model_executor/models/interfaces_base.py
+++ b/vllm/model_executor/models/interfaces_base.py
@@ -1,7 +1,8 @@
 # SPDX-License-Identifier: Apache-2.0
 # SPDX-FileCopyrightText: Copyright contributors to the vLLM project
-from typing import (TYPE_CHECKING, Any, ClassVar, Literal, Optional, Protocol,
-                    Union, overload, runtime_checkable)
+
+from typing import (TYPE_CHECKING, Optional, Protocol, Union, overload,
+                    runtime_checkable)
 
 import torch
 import torch.nn as nn
@@ -12,10 +13,9 @@ from vllm.utils import supports_kw
 
 if TYPE_CHECKING:
     from vllm.config import VllmConfig
-    from vllm.model_executor.layers.pooler import Pooler
-else:
-    VllmConfig = Any
-    Pooler = Any
+    from vllm.model_executor.layers.pooler import PoolerOutput
+    from vllm.model_executor.pooling_metadata import PoolingMetadata
+    from vllm.model_executor.sampling_metadata import SamplingMetadata
 
 logger = init_logger(__name__)
 
@@ -36,7 +36,7 @@ class VllmModel(Protocol[T_co]):
 
     def __init__(
         self,
-        vllm_config: VllmConfig,
+        vllm_config: "VllmConfig",
         prefix: str = "",
     ) -> None:
         ...
@@ -98,6 +98,7 @@ class VllmModelForTextGeneration(VllmModel[T], Protocol[T]):
     def compute_logits(
         self,
         hidden_states: T,
+        sampling_metadata: "SamplingMetadata",
     ) -> Optional[T]:
         """Return `None` if TP rank > 0."""
         ...
@@ -129,31 +130,16 @@ def is_text_generation_model(
 
 
 @runtime_checkable
-class VllmModelForPooling(VllmModel[T_co], Protocol[T_co]):
+class VllmModelForPooling(VllmModel[T], Protocol[T]):
     """The interface required for all pooling models in vLLM."""
 
-    is_pooling_model: ClassVar[Literal[True]] = True
-    """
-    A flag that indicates this model supports pooling.
-
-    Note:
-        There is no need to redefine this flag if this class is in the
-        MRO of your model class.
-    """
-
-    default_pooling_type: ClassVar[str] = "LAST"
-    """
-    Indicates the
-    [vllm.model_executor.layers.pooler.PoolerConfig.pooling_type][]
-    to use by default.
-
-    You can use the
-    [vllm.model_executor.models.interfaces_base.default_pooling_type][]
-    decorator to conveniently set this field.
-    """
-
-    pooler: Pooler
-    """The pooler is only called on TP rank 0."""
+    def pooler(
+        self,
+        hidden_states: T,
+        pooling_metadata: "PoolingMetadata",
+    ) -> "PoolerOutput":
+        """Only called on TP rank 0."""
+        ...
 
 
 @overload
@@ -172,21 +158,7 @@ def is_pooling_model(
     if not is_vllm_model(model):
         return False
 
-    return getattr(model, "is_pooling_model", False)
-
-
-_T = TypeVar("_T", bound=type[nn.Module])
-
-
-def default_pooling_type(pooling_type: str):
-    """Decorator to set `VllmModelForPooling.default_pooling_type`."""
-
-    def func(model: _T) -> _T:
-        model.default_pooling_type = pooling_type  # type: ignore
-        return model
-
-    return func
-
+    if isinstance(model, type):
+        return isinstance(model, VllmModelForPooling)
 
-def get_default_pooling_type(model: Union[type[object], object]) -> str:
-    return getattr(model, "default_pooling_type", "LAST")
+    return isinstance(model, VllmModelForPooling)

