diff --git a/vllm/model_executor/models/mlp_speculator.py b/vllm/model_executor/models/mlp_speculator.py
index d057eb49a..c6a97388d 100644
--- a/vllm/model_executor/models/mlp_speculator.py
+++ b/vllm/model_executor/models/mlp_speculator.py
@@ -8,7 +8,9 @@ import torch
 import torch.nn as nn
 
 from vllm.config import VllmConfig
+from vllm.model_executor import SamplingMetadata
 from vllm.model_executor.layers.logits_processor import LogitsProcessor
+from vllm.model_executor.layers.sampler import SamplerOutput, get_sampler
 from vllm.model_executor.layers.vocab_parallel_embedding import (
     ParallelLMHead, VocabParallelEmbedding)
 from vllm.model_executor.model_loader.weight_utils import default_weight_loader
@@ -139,57 +141,55 @@ class MLPSpeculator(nn.Module):
         self.config = config
         self.logits_processor = LogitsProcessor(config.vocab_size,
                                                 config.vocab_size, 1.0)
+        self.sampler = get_sampler()
 
-    # NOTE(woosuk): This method is commented out because it is old code
-    # using V0. We should either port it to V1 or remove it.
-
-    # def generate_proposals(
-    #     self,
-    #     input_ids: torch.Tensor,
-    #     previous_hidden_states: torch.Tensor,
-    #     num_predict_tokens: int,
-    #     sampling_metadata: SamplingMetadata,
-    # ) -> list[SamplerOutput]:
-    #     if num_predict_tokens > self.max_speculative_tokens:
-    #         raise ValueError(f"Max speculative tokens for model is "
-    #                          f"{self.max_speculative_tokens}, but "
-    #                          f"{num_predict_tokens} were requested")
-
-    #     # b x 1 x d
-    #     previous_hidden_states = previous_hidden_states.unsqueeze(1)
+    def generate_proposals(
+        self,
+        input_ids: torch.Tensor,
+        previous_hidden_states: torch.Tensor,
+        num_predict_tokens: int,
+        sampling_metadata: SamplingMetadata,
+    ) -> list[SamplerOutput]:
+        if num_predict_tokens > self.max_speculative_tokens:
+            raise ValueError(f"Max speculative tokens for model is "
+                             f"{self.max_speculative_tokens}, but "
+                             f"{num_predict_tokens} were requested")
+
+        # b x 1 x d
+        previous_hidden_states = previous_hidden_states.unsqueeze(1)
 
-    #     if self.scale_input:
-    #         previous_hidden_states = self.ln0(previous_hidden_states) / SQRT2
+        if self.scale_input:
+            previous_hidden_states = self.ln0(previous_hidden_states) / SQRT2
 
-    #     # b x 1
-    #     last_tokens = input_ids.unsqueeze(1)
+        # b x 1
+        last_tokens = input_ids.unsqueeze(1)
 
-    #     next_tokens = []
+        next_tokens = []
 
-    #     for head_index in range(num_predict_tokens):
+        for head_index in range(num_predict_tokens):
 
-    #         # Project and predict
-    #         z = self.emb[head_index](last_tokens)  # b k d
-    #         states = self.proj[head_index](previous_hidden_states)
+            # Project and predict
+            z = self.emb[head_index](last_tokens)  # b k d
+            states = self.proj[head_index](previous_hidden_states)
 
-    #         # Weighted add of state_weight*state and emb_weight*z
-    #         # Let subsequent LN take care of denominator
-    #         # state_weight is close to 1, so shouldn't be any precision issues
-    #         states.add_(z, alpha=self.emb_weight / self.state_weight)
+            # Weighted add of state_weight*state and emb_weight*z
+            # Let subsequent LN take care of denominator
+            # state_weight is close to 1, so shouldn't be any precision issues
+            states.add_(z, alpha=self.emb_weight / self.state_weight)
 
-    #         states = self.activation(self.ln[head_index](states))  # b k d
-    #         previous_hidden_states = states
-    #         # TODO: not yet supporting top_k_tokens_per_head
-    #         states = states.flatten(0, 1)
+            states = self.activation(self.ln[head_index](states))  # b k d
+            previous_hidden_states = states
+            # TODO: not yet supporting top_k_tokens_per_head
+            states = states.flatten(0, 1)
 
-    #         logits = self.logits_processor(self.head[head_index], states,
-    #                                        sampling_metadata)
+            logits = self.logits_processor(self.head[head_index], states,
+                                           sampling_metadata)
 
-    #         output = self.sampler(logits, sampling_metadata)
-    #         last_tokens = output.sampled_token_ids
-    #         next_tokens.append(output)
+            output = self.sampler(logits, sampling_metadata)
+            last_tokens = output.sampled_token_ids
+            next_tokens.append(output)
 
-    #     return next_tokens
+        return next_tokens
 
     def load_weights(self, weights: Iterable[tuple[str,
                                                    torch.Tensor]]) -> set[str]:

