diff --git a/vllm/v1/core/sched/interface.py b/vllm/v1/core/sched/interface.py
index 5b1de3a66..dd5052a34 100644
--- a/vllm/v1/core/sched/interface.py
+++ b/vllm/v1/core/sched/interface.py
@@ -9,7 +9,7 @@ if TYPE_CHECKING:
     from vllm.v1.core.sched.output import SchedulerOutput
     from vllm.v1.engine import EngineCoreOutputs
     from vllm.v1.metrics.stats import SchedulerStats
-    from vllm.v1.outputs import DraftTokenIds, ModelRunnerOutput
+    from vllm.v1.outputs import ModelRunnerOutput
     from vllm.v1.request import Request, RequestStatus
 
 
@@ -61,14 +61,6 @@ class SchedulerInterface(ABC):
         """
         raise NotImplementedError
 
-    @abstractmethod
-    def update_draft_token_ids(
-        self,
-        draft_token_ids: "DraftTokenIds",
-    ) -> None:
-        """Update the draft token ids for the scheduled requests."""
-        raise NotImplementedError
-
     @abstractmethod
     def add_request(self, request: "Request") -> None:
         """Add a new request to the scheduler's internal queue.

