diff --git a/csrc/quantization/per_token_group_quant_8bit.h b/csrc/quantization/per_token_group_quant_8bit.h
deleted file mode 100644
index 537b61bc4..000000000
--- a/csrc/quantization/per_token_group_quant_8bit.h
+++ /dev/null
@@ -1,10 +0,0 @@
-#pragma once
-#include <torch/all.h>
-
-// TODO(wentao): refactor the folder to 8bit, then includes fp8 and int8 folders
-// 8-bit per-token-group quantization helper used by both FP8 and INT8
-void per_token_group_quant_8bit(const torch::Tensor& input,
-                                torch::Tensor& output_q,
-                                torch::Tensor& output_s, int64_t group_size,
-                                double eps, double min_8bit, double max_8bit,
-                                bool scale_ue8m0 = false);
\ No newline at end of file

