diff --git a/vllm/compilation/fix_functionalization.py b/vllm/compilation/fix_functionalization.py
index 54403c1f7..286221d32 100644
--- a/vllm/compilation/fix_functionalization.py
+++ b/vllm/compilation/fix_functionalization.py
@@ -9,7 +9,6 @@ import torch
 from torch._higher_order_ops.auto_functionalize import auto_functionalized
 
 from vllm.logger import init_logger
-from vllm.platforms import current_platform
 
 from .fx_utils import is_func
 from .vllm_inductor_pass import VllmInductorPass
@@ -26,14 +25,9 @@ class FixFunctionalizationPass(VllmInductorPass):
     To add new nodes to defunctionalize, add to the if-elif chain in __call__.
     """
 
-    @VllmInductorPass.time_and_log
     def __call__(self, graph: torch.fx.Graph):
-        # XPU does not support auto-functionalization yet.
-        # Will enable this when switch to vllm-xpu-kernels.
-        if current_platform.is_xpu():
-            logger.debug("XPU platform does not support fix functionalization"
-                         "pass currently.")
-            return
+        self.begin()
+        self.dump_graph(graph, "before_fix_functionalization")
 
         self.nodes_to_remove: list[torch.fx.Node] = []
         count = 0
@@ -95,21 +89,12 @@ class FixFunctionalizationPass(VllmInductorPass):
                                      node,
                                      mutated_args,
                                      args=('result', 'input', 'scale'))
-            elif hasattr(
-                    torch.ops._C, "silu_and_mul_nvfp4_quant"
-            ) and at_target == torch.ops._C.silu_and_mul_nvfp4_quant.default:
-                mutated_args = {1: 'result', 2: 'result_block_scale'}
-                self.defunctionalize(graph,
-                                     node,
-                                     mutated_args,
-                                     args=('result', 'result_block_scale',
-                                           'input', 'input_global_scale'))
             else:
                 continue  # skip the count
 
             count += 1
 
-        self.dump_graph(graph, "before_cleanup")
+        self.dump_graph(graph, "before_fix_functionalization_cleanup")
 
         # Remove the nodes all at once
         count_removed = len(self.nodes_to_remove)
@@ -118,7 +103,8 @@ class FixFunctionalizationPass(VllmInductorPass):
 
         logger.debug("De-functionalized %s nodes, removed %s nodes", count,
                      count_removed)
-        self.nodes_to_remove.clear()
+        self.dump_graph(graph, "after_fix_functionalization")
+        self.end_and_log()
 
     def _remove(self, node_or_nodes: Union[torch.fx.Node,
                                            Iterable[torch.fx.Node]]):

