diff --git a/vllm/compilation/fx_utils.py b/vllm/compilation/fx_utils.py
index 2db8b5441..9ef388932 100644
--- a/vllm/compilation/fx_utils.py
+++ b/vllm/compilation/fx_utils.py
@@ -2,7 +2,7 @@
 # SPDX-FileCopyrightText: Copyright contributors to the vLLM project
 
 import operator
-from collections.abc import Iterable, Iterator
+from collections.abc import Iterable
 from typing import Optional
 
 from torch import fx
@@ -14,10 +14,6 @@ def is_func(node: fx.Node, target) -> bool:
     return node.op == "call_function" and node.target == target
 
 
-def is_auto_func(node: fx.Node, op: OpOverload) -> bool:
-    return is_func(node, auto_functionalized) and node.args[0] == op
-
-
 # Returns the first specified node with the given op (if it exists)
 def find_specified_fn_maybe(nodes: Iterable[fx.Node],
                             op: OpOverload) -> Optional[fx.Node]:
@@ -64,21 +60,3 @@ def find_getitem(node: fx.Node, idx: int) -> fx.Node:
     ret = find_getitem_maybe(node, idx)
     assert ret is not None, f"Could not find getitem {idx} in node {node}"
     return ret
-
-
-# An auto-functionalization-aware utility for finding nodes with a specific op
-def find_op_nodes(op: OpOverload, graph: fx.Graph) -> Iterator[fx.Node]:
-    if not op._schema.is_mutable:
-        yield from graph.find_nodes(op="call_function", target=op)
-
-    for n in graph.find_nodes(op="call_function", target=auto_functionalized):
-        if n.args[0] == op:
-            yield n
-
-
-# Asserts that the node only has one user and returns it
-# Even if a node has only 1 user, it might share storage with another node,
-# which might need to be taken into account.
-def get_only_user(node: fx.Node) -> fx.Node:
-    assert len(node.users) == 1
-    return next(iter(node.users))

