diff --git a/vllm/compilation/inductor_pass.py b/vllm/compilation/inductor_pass.py
index e1b691df3..810d0801e 100644
--- a/vllm/compilation/inductor_pass.py
+++ b/vllm/compilation/inductor_pass.py
@@ -1,7 +1,6 @@
 # SPDX-License-Identifier: Apache-2.0
 # SPDX-FileCopyrightText: Copyright contributors to the vLLM project
 
-import functools
 import hashlib
 import inspect
 import json
@@ -11,8 +10,6 @@ from typing import Any, Callable, Optional, Union
 
 import torch
 from torch import fx
-from torch._subclasses.fake_tensor import (FakeTensorMode,
-                                           unset_fake_temporarily)
 
 from vllm.utils import is_torch_equal_or_newer
 
@@ -79,10 +76,9 @@ class InductorPass(CustomGraphPass):
         for src in srcs:
             if isinstance(src, str):
                 src_str = src
-            elif isinstance(src, (types.FunctionType, type)):
+            elif isinstance(src, types.FunctionType):
                 src_str = inspect.getsource(src)
             else:
-                # object instance
                 src_str = inspect.getsource(src.__class__)
             hasher.update(src_str.encode("utf-8"))
         return hasher.hexdigest()
@@ -117,20 +113,3 @@ class CallableInductorPass(InductorPass):
 
     def uuid(self) -> Any:
         return self._uuid
-
-
-def enable_fake_mode(fn: Callable[..., Any]) -> Callable[..., Any]:
-    """
-    Applies a FakeTensorMode context. This is useful when you don't want to
-    create or run things with real tensors.
-    """
-
-    @functools.wraps(fn)
-    def fn_new(*args, **kwargs) -> Any:
-        with torch._guards.tracing(
-                None), unset_fake_temporarily(), FakeTensorMode():
-            result = fn(*args, **kwargs)
-
-        return result
-
-    return fn_new

