diff --git a/vllm/compilation/wrapper.py b/vllm/compilation/wrapper.py
index 930e4d27b..2a261c84c 100644
--- a/vllm/compilation/wrapper.py
+++ b/vllm/compilation/wrapper.py
@@ -10,8 +10,8 @@ from typing import Callable, Optional
 
 import torch
 
-from vllm.config import (CompilationLevel, CUDAGraphMode,
-                         get_current_vllm_config)
+import vllm.envs as envs
+from vllm.config import CompilationLevel, get_current_vllm_config
 from vllm.logger import init_logger
 
 logger = init_logger(__name__)
@@ -46,10 +46,11 @@ class TorchCompileWrapperWithCustomDispatcher:
                 options = get_current_vllm_config(
                 ).compilation_config.inductor_compile_config
 
-            compiled_callable = torch.compile(self.forward,
-                                              fullgraph=True,
-                                              backend=backend,
-                                              options=options)
+            compiled_callable = torch.compile(
+                self.forward,
+                fullgraph=envs.VLLM_TEST_DYNAMO_FULLGRAPH_CAPTURE,
+                backend=backend,
+                options=options)
 
         self.compiled_callable = compiled_callable
         self.original_code_object = self.__class__.forward.__code__
@@ -92,10 +93,9 @@ class TorchCompileWrapperWithCustomDispatcher:
             return
 
         self.compiled_codes.append(new_code)
-        debug_dump_dir = self.vllm_config.compilation_config.debug_dump_path
-        if isinstance(debug_dump_dir, str) and debug_dump_dir != "":
-            rank = self.vllm_config.parallel_config.rank
-            decompiled_file = os.path.join(debug_dump_dir, f"rank_{rank}",
+        local_cache_dir = self.vllm_config.compilation_config.local_cache_dir
+        if isinstance(local_cache_dir, str):
+            decompiled_file = os.path.join(local_cache_dir,
                                            "transformed_code.py")
             if not os.path.exists(decompiled_file):
                 try:
@@ -105,7 +105,6 @@ class TorchCompileWrapperWithCustomDispatcher:
                     # not a reversible process.
                     import depyf
                     src = depyf.decompile(new_code)
-
                     with open(decompiled_file, "w") as f:
                         f.write(src)
 
@@ -114,8 +113,8 @@ class TorchCompileWrapperWithCustomDispatcher:
                 except Exception:
                     pass
 
-        if self.vllm_config.compilation_config.cudagraph_mode != \
-            CUDAGraphMode.NONE and "update" in new_code.co_names:
+        if self.vllm_config.compilation_config.use_cudagraph and \
+            "update" in new_code.co_names:
             import depyf
             src = depyf.decompile(new_code)
             msg = "Assigning / modifying buffers of nn.Module during forward pass is not allowed when using cudagraph inside the compiler because it will cause silent errors. Please use eager mode or fix the code. The following code contains clues about which buffer is being modified (please search for the usage of the function `update`):\n" + src  # noqa

