diff --git a/flagscale/runner/runner_inference.py b/flagscale/runner/runner_inference.py
index aba0c56c..7eb850bc 100644
--- a/flagscale/runner/runner_inference.py
+++ b/flagscale/runner/runner_inference.py
@@ -20,7 +20,10 @@ from flagscale.runner.utils import (
 
 def _get_args_vllm(config: DictConfig):
     # step1: yaml -> dict
-    assert config.experiment.task.backend in ["vllm"], "This function only supports vllm backend."
+    assert config.experiment.task.backend in [
+        "vllm",
+        "lerobot",
+    ], "This function only supports vllm/lerobot backend."
     config_dict = OmegaConf.to_container(config, resolve=True)
 
     # step2: restructuring the config

