diff --git a/megatron/bridge/utils/path_utils.py b/megatron/bridge/utils/path_utils.py
new file mode 100644
index 000000000..bcfcabc0a
--- /dev/null
+++ b/megatron/bridge/utils/path_utils.py
@@ -0,0 +1,19 @@
+# Copyright (c) 2025, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific
+
+from pathlib import Path
+
+
+def resolve_path(path: str) -> Path:
+    """Resolve a path to an absolute path."""
+    return Path(path).expanduser().absolute().resolve()
\ No newline at end of file
