diff --git a/megatron-lm-musa-patch/examples/llama2/stop_all.sh b/megatron-lm-musa-patch/examples/llama2/stop_all.sh
new file mode 100644
index 00000000..d64eaa4b
--- /dev/null
+++ b/megatron-lm-musa-patch/examples/llama2/stop_all.sh
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+HOSTFILE=./hostfile
+NUM_NODES=$(grep -v '^#\|^$' $HOSTFILE | wc -l)
+echo "NUM_NODES: $NUM_NODES"
+
+hostlist=$(grep -v '^#\|^$' $HOSTFILE | awk '{print $1}' | xargs)
+for host in ${hostlist[@]}; do
+    ssh -f -n $host "pkill -f '/opt/conda/envs/py38/bin/torchrun'"
+    ssh -f -n $host "pkill -f '/usr/local/bin/torchrun'"
+    echo "$host is killed."
+done

