diff --git a/vllm/distributed/kv_transfer/kv_connector/v1/multi_connector.py b/vllm/distributed/kv_transfer/kv_connector/v1/multi_connector.py
index 6836a71e5..be3c23399 100644
--- a/vllm/distributed/kv_transfer/kv_connector/v1/multi_connector.py
+++ b/vllm/distributed/kv_transfer/kv_connector/v1/multi_connector.py
@@ -1,29 +1,23 @@
 # SPDX-License-Identifier: Apache-2.0
 # SPDX-FileCopyrightText: Copyright contributors to the vLLM project
 import copy
-from collections.abc import Iterable
 from dataclasses import dataclass
 from typing import TYPE_CHECKING, Any, Optional
 
 import torch
 
-from vllm.config import VllmConfig
-from vllm.config.kv_transfer import KVTransferConfig
+from vllm.config import KVTransferConfig, VllmConfig
 from vllm.distributed.kv_transfer.kv_connector.factory import (
     KVConnectorFactory)
 from vllm.distributed.kv_transfer.kv_connector.v1.base import (
     KVConnectorBase_V1, KVConnectorMetadata, KVConnectorRole)
-from vllm.distributed.kv_transfer.kv_connector.v1.metrics import (
-    KVConnectorStats)
 from vllm.logger import init_logger
+from vllm.v1.core.kv_cache_manager import KVCacheBlocks
 from vllm.v1.core.sched.output import SchedulerOutput
-from vllm.v1.outputs import KVConnectorOutput
 
 if TYPE_CHECKING:
     from vllm.attention.backends.abstract import AttentionMetadata
-    from vllm.distributed.kv_events import KVCacheEvent
     from vllm.forward_context import ForwardContext
-    from vllm.v1.core.kv_cache_manager import KVCacheBlocks
     from vllm.v1.request import Request
 
 logger = init_logger(__name__)
@@ -35,43 +29,6 @@ class MultiKVConnectorMetadata(KVConnectorMetadata):
     extra_async_saves: Optional[dict[str, int]] = None
 
 
-@dataclass
-class MultiKVConnectorStats(KVConnectorStats):
-    """
-    Maintain a dict of KVConnectorStats objects, one for each connector.
-    This is used to aggregate the stats from all connectors separately.
-    """
-
-    def aggregate(self, other: KVConnectorStats) -> KVConnectorStats:
-        for connector_id, stats in other.data.items():
-            if connector_id not in self.data:
-                self[connector_id] = stats
-            else:
-                assert isinstance(stats, type(self.data[connector_id]))
-                self[connector_id] = self[connector_id].aggregate(stats)
-        return self
-
-    def reset(self):
-        for stats in self.data.values():
-            stats.reset()
-
-    def reduce(self) -> dict[str, Any]:
-        # TODO (NickLucche) Adjust for logging on separate lines
-        return {
-            connector_id: stats.reduce()
-            for connector_id, stats in self.data.items()
-        }
-
-    def is_empty(self) -> bool:
-        return all(stats.is_empty() for stats in self.data.values())
-
-    def __getitem__(self, connector_id: str) -> KVConnectorStats:
-        return self.data[connector_id]
-
-    def __setitem__(self, connector_id: str, stats: KVConnectorStats):
-        self.data[connector_id] = stats
-
-
 class MultiConnector(KVConnectorBase_V1):
     """
     A wrapper for using multiple KVConnectors at the same time.
@@ -85,19 +42,14 @@ class MultiConnector(KVConnectorBase_V1):
     def __init__(self, vllm_config: "VllmConfig", role: KVConnectorRole):
         super().__init__(vllm_config=vllm_config, role=role)
         self._connectors: list[KVConnectorBase_V1] = []
-        self._ktc_kv_transfer_config = []
         ktcs = vllm_config.kv_transfer_config.kv_connector_extra_config.get(
             "connectors")
         assert ktcs is not None
         for ktc in ktcs:
             temp_config = copy.copy(vllm_config)
-            engine_id = ktc.get("engine_id",
-                                vllm_config.kv_transfer_config.engine_id)
-            temp_config.kv_transfer_config = KVTransferConfig(
-                **ktc, engine_id=engine_id)
+            temp_config.kv_transfer_config = KVTransferConfig(**ktc)
             self._connectors.append(
-                KVConnectorFactory.create_connector(temp_config, role))
-            self._ktc_kv_transfer_config.append(temp_config.kv_transfer_config)
+                KVConnectorFactory.create_connector_v1(temp_config, role))
 
         # A mapping from request id to the index of the connector chosen to
         # load the request from (if any).
@@ -129,18 +81,6 @@ class MultiConnector(KVConnectorBase_V1):
         for c in self._connectors:
             c.clear_connector_metadata()
 
-    def shutdown(self):
-        exception: Optional[Exception] = None
-        for c in self._connectors:
-            try:
-                c.shutdown()
-            except Exception as e:
-                logger.exception("Exception during connector %s shutdown.",
-                                 c.__class__.__name__)
-                exception = e
-        if exception:
-            raise exception
-
     # ==============================
     # Worker-side methods
     # ==============================
@@ -196,15 +136,11 @@ class MultiConnector(KVConnectorBase_V1):
         self,
         request: "Request",
         num_computed_tokens: int,
-    ) -> tuple[Optional[int], bool]:
+    ) -> tuple[int, bool]:
         to_return = (0, False)
         for i, c in enumerate(self._connectors):
             toks, load_async = c.get_num_new_matched_tokens(
                 request, num_computed_tokens)
-            # If there is a connector still looking up the matches,
-            # we return None to indicate that we are not done yet.
-            if toks is None:
-                return (None, False)
             # The first connector that has new matched tokens will be assigned
             # to this request.
             if to_return[0] == 0 and toks > 0:
@@ -238,10 +174,6 @@ class MultiConnector(KVConnectorBase_V1):
             self._extra_async_saves = {}
         return metadata
 
-    def update_connector_output(self, connector_output: KVConnectorOutput):
-        for c in self._connectors:
-            c.update_connector_output(connector_output)
-
     def request_finished(
         self,
         request: "Request",
@@ -255,7 +187,7 @@ class MultiConnector(KVConnectorBase_V1):
                 async_saves += 1
             if txfer_params is not None:
                 if kv_txfer_params is not None:
-                    # TODO we can probably change this to merge the dicts here,
+                    #TODO we can probably change this to merge the dicts here,
                     # checking for key clashes.
                     raise RuntimeError(
                         "Only one connector can produce KV transfer params")
@@ -267,62 +199,3 @@ class MultiConnector(KVConnectorBase_V1):
         self._requests_to_connector.pop(request.request_id, None)
 
         return async_saves > 0, kv_txfer_params
-
-    def take_events(self) -> Iterable["KVCacheEvent"]:
-        for c in self._connectors:
-            yield from c.take_events()
-
-    @classmethod
-    def get_required_kvcache_layout(
-            cls, vllm_config: "VllmConfig") -> Optional[str]:
-        """
-        Get the required KV cache layout for this connector.
-        Args:
-            vllm_config (VllmConfig): the vllm config.
-
-        Returns:
-            str: the required KV cache layout. e.g. HND, or NHD.
-            None if the connector does not require a specific layout.
-        """
-        ktcs = vllm_config.kv_transfer_config.kv_connector_extra_config.get(
-            "connectors")
-        assert ktcs is not None
-        layouts: set[str] = set()
-        temp_vllm_config = copy.copy(vllm_config)
-        for ktc in ktcs:
-            kv_transfer_config = KVTransferConfig(**ktc)
-            temp_vllm_config.kv_transfer_config = kv_transfer_config
-            connector_cls = KVConnectorFactory.get_connector_class(
-                kv_transfer_config)
-            required_kvcache_layout = (
-                connector_cls.get_required_kvcache_layout(temp_vllm_config))
-            if required_kvcache_layout is not None:
-                layouts.add(required_kvcache_layout)
-
-        if len(layouts) > 1:
-            raise ValueError(f"KV cache layout mismatch: "
-                             f"found {len(layouts)} different layouts "
-                             f"({', '.join(layouts) })."
-                             f"All connectors must use the same layout.")
-        return next(iter(layouts), None)
-
-    @classmethod
-    def build_kv_connector_stats(
-            cls,
-            data: Optional[dict[str,
-                                Any]] = None) -> Optional[KVConnectorStats]:
-        return MultiKVConnectorStats(data=data) if data is not None \
-            else MultiKVConnectorStats()
-
-    def get_kv_connector_stats(self) -> Optional[MultiKVConnectorStats]:
-        # Group connector stats by connector type.
-        stats_by_connector: Optional[MultiKVConnectorStats] = None
-        for c in self._connectors:
-            stats = c.get_kv_connector_stats()
-            if stats is None:
-                continue
-            if stats_by_connector is None:
-                # Lazy init to allow optional return value.
-                stats_by_connector = MultiKVConnectorStats()
-            stats_by_connector[c.__class__.__name__] = stats
-        return stats_by_connector

