diff --git a/csrc/quantization/machete/generate.py b/csrc/quantization/machete/generate.py
index 8fd536ef4..9af7833d0 100644
--- a/csrc/quantization/machete/generate.py
+++ b/csrc/quantization/machete/generate.py
@@ -349,12 +349,9 @@ def to_cute_constant(value: list[int]):
 
 
 def unique_schedules(impl_configs: list[ImplConfig]):
-    # Use dict over set for deterministic ordering
-    return list({
-        sch: None
-        for impl_config in impl_configs
-        for sch in impl_config.schedules
-    }.keys())
+    return list(
+        set(sch for impl_config in impl_configs
+            for sch in impl_config.schedules))
 
 
 def unsigned_type_with_bitwidth(num_bits):
@@ -417,7 +414,7 @@ def create_sources(impl_configs: list[ImplConfig], num_impl_files=8):
             ))
 
     def prepacked_type_key(prepack_type: PrepackTypeConfig):
-        # For now, we can just use the first accumulator type seen since
+        # For now we we can just use the first accumulator type seen since
         # the tensor core shapes/layouts don't vary based on accumulator
         # type so we can generate less code this way
         return (prepack_type.a, prepack_type.b_num_bits, prepack_type.convert)
@@ -571,79 +568,78 @@ def generate():
                      itertools.repeat(default_heuristic))
     ]
 
-    # TODO: Support W4A8 when ready
-    # # Stored as "condition": ((tile_shape_mn), (cluster_shape_mnk))
-    # # TODO (LucasWilkinson): Further tuning required
-    # qqq_tile_heuristic_config = {
-    #     #### M = 257+
-    #     # ((128, 256), (2, 1, 1)) Broken for QQQ types
-    #     # TODO (LucasWilkinson): Investigate further
-    #     # "M > 256 && K <= 16384 && N <= 4096": ((128, 128), (2, 1, 1)),
-    #     # "M > 256": ((128, 256), (2, 1, 1)),
-    #     "M > 256": ((128, 128), (2, 1, 1)),
-    #     #### M = 129-256
-    #     "M > 128 && K <= 4096 && N <= 4096": ((128, 64), (2, 1, 1)),
-    #     "M > 128 && K <= 8192 && N <= 8192": ((128, 128), (2, 1, 1)),
-    #     # ((128, 256), (2, 1, 1)) Broken for QQQ types
-    #     # TODO (LucasWilkinson): Investigate further
-    #     # "M > 128": ((128, 256), (2, 1, 1)),
-    #     "M > 128": ((128, 128), (2, 1, 1)),
-    #     #### M = 65-128
-    #     "M > 64 && K <= 4069 && N <= 4069": ((128, 32), (2, 1, 1)),
-    #     "M > 64 && K <= 4069 && N <= 8192": ((128, 64), (2, 1, 1)),
-    #     "M > 64 && K >= 8192 && N >= 12288": ((256, 128), (2, 1, 1)),
-    #     "M > 64": ((128, 128), (2, 1, 1)),
-    #     #### M = 33-64
-    #     "M > 32 && K <= 6144 && N <= 6144": ((128, 16), (1, 1, 1)),
-    #     # Broken for QQQ types
-    #     # TODO (LucasWilkinson): Investigate further
-    #     #"M > 32 && K >= 16384 && N >= 12288": ((256, 64), (2, 1, 1)),
-    #     "M > 32": ((128, 64), (2, 1, 1)),
-    #     #### M = 17-32
-    #     "M > 16 && K <= 12288 && N <= 8192": ((128, 32), (2, 1, 1)),
-    #     "M > 16": ((256, 32), (2, 1, 1)),
-    #     #### M = 1-16
-    #     "N >= 26624": ((256, 16), (1, 1, 1)),
-    #     None: ((128, 16), (1, 1, 1)),
-    # }
-
-    # # For now we use the same heuristic for all types
-    # # Heuristic is currently tuned for H100s
-    # qqq_heuristic = [
-    #     (cond, ScheduleConfig(*tile_config,
-    #                           **sch_common_params))  # type: ignore
-    #     for cond, tile_config in qqq_tile_heuristic_config.items()
-    # ]
-
-    # QQQ_kernel_types = [
-    #     *(TypeConfig(
-    #         a=DataType.s8,
-    #         b=VLLMDataType.u4b8,
-    #         b_group_scale=b_group_scale,
-    #         b_group_zeropoint=DataType.void,
-    #         b_channel_scale=DataType.f32,
-    #         a_token_scale=DataType.f32,
-    #         out=DataType.f16,
-    #         accumulator=DataType.s32,
-    #     ) for b_group_scale in (DataType.f16, DataType.void)),
-    #     *(TypeConfig(
-    #         a=DataType.e4m3,
-    #         b=VLLMDataType.u4b8,
-    #         b_group_scale=b_group_scale,
-    #         b_group_zeropoint=DataType.void,
-    #         b_channel_scale=DataType.f32,
-    #         a_token_scale=DataType.f32,
-    #         out=DataType.f16,
-    #         accumulator=DataType.f32,
-    #     ) for b_group_scale in (DataType.f16, DataType.void)),
-    # ]
-
-    # impl_configs += [
-    #     ImplConfig(x[0], x[1], x[2])
-    #     for x in zip(QQQ_kernel_types,
-    #                  itertools.repeat(get_unique_schedules(qqq_heuristic)),
-    #                  itertools.repeat(qqq_heuristic))
-    # ]
+    # Stored as "condition": ((tile_shape_mn), (cluster_shape_mnk))
+    # TODO (LucasWilkinson): Further tuning required
+    qqq_tile_heuristic_config = {
+        #### M = 257+
+        # ((128, 256), (2, 1, 1)) Broken for QQQ types
+        # TODO (LucasWilkinson): Investigate further
+        # "M > 256 && K <= 16384 && N <= 4096": ((128, 128), (2, 1, 1)),
+        # "M > 256": ((128, 256), (2, 1, 1)),
+        "M > 256": ((128, 128), (2, 1, 1)),
+        #### M = 129-256
+        "M > 128 && K <= 4096 && N <= 4096": ((128, 64), (2, 1, 1)),
+        "M > 128 && K <= 8192 && N <= 8192": ((128, 128), (2, 1, 1)),
+        # ((128, 256), (2, 1, 1)) Broken for QQQ types
+        # TODO (LucasWilkinson): Investigate further
+        # "M > 128": ((128, 256), (2, 1, 1)),
+        "M > 128": ((128, 128), (2, 1, 1)),
+        #### M = 65-128
+        "M > 64 && K <= 4069 && N <= 4069": ((128, 32), (2, 1, 1)),
+        "M > 64 && K <= 4069 && N <= 8192": ((128, 64), (2, 1, 1)),
+        "M > 64 && K >= 8192 && N >= 12288": ((256, 128), (2, 1, 1)),
+        "M > 64": ((128, 128), (2, 1, 1)),
+        #### M = 33-64
+        "M > 32 && K <= 6144 && N <= 6144": ((128, 16), (1, 1, 1)),
+        # Broken for QQQ types
+        # TODO (LucasWilkinson): Investigate further
+        #"M > 32 && K >= 16384 && N >= 12288": ((256, 64), (2, 1, 1)),
+        "M > 32": ((128, 64), (2, 1, 1)),
+        #### M = 17-32
+        "M > 16 && K <= 12288 && N <= 8192": ((128, 32), (2, 1, 1)),
+        "M > 16": ((256, 32), (2, 1, 1)),
+        #### M = 1-16
+        "N >= 26624": ((256, 16), (1, 1, 1)),
+        None: ((128, 16), (1, 1, 1)),
+    }
+
+    # For now we use the same heuristic for all types
+    # Heuristic is currently tuned for H100s
+    qqq_heuristic = [
+        (cond, ScheduleConfig(*tile_config,
+                              **sch_common_params))  # type: ignore
+        for cond, tile_config in qqq_tile_heuristic_config.items()
+    ]
+
+    QQQ_kernel_types = [
+        *(TypeConfig(
+            a=DataType.s8,
+            b=VLLMDataType.u4b8,
+            b_group_scale=b_group_scale,
+            b_group_zeropoint=DataType.void,
+            b_channel_scale=DataType.f32,
+            a_token_scale=DataType.f32,
+            out=DataType.f16,
+            accumulator=DataType.s32,
+        ) for b_group_scale in (DataType.f16, DataType.void)),
+        *(TypeConfig(
+            a=DataType.e4m3,
+            b=VLLMDataType.u4b8,
+            b_group_scale=b_group_scale,
+            b_group_zeropoint=DataType.void,
+            b_channel_scale=DataType.f32,
+            a_token_scale=DataType.f32,
+            out=DataType.f16,
+            accumulator=DataType.f32,
+        ) for b_group_scale in (DataType.f16, DataType.void)),
+    ]
+
+    impl_configs += [
+        ImplConfig(x[0], x[1], x[2])
+        for x in zip(QQQ_kernel_types,
+                     itertools.repeat(get_unique_schedules(qqq_heuristic)),
+                     itertools.repeat(qqq_heuristic))
+    ]
 
     output_dir = os.path.join(SCRIPT_DIR, "generated")
 

