diff --git a/vllm/model_executor/layers/quantization/kernels/mixed_precision/bitblas.py b/vllm/model_executor/layers/quantization/kernels/mixed_precision/bitblas.py
index 0eca3b4c0..649d07b4d 100644
--- a/vllm/model_executor/layers/quantization/kernels/mixed_precision/bitblas.py
+++ b/vllm/model_executor/layers/quantization/kernels/mixed_precision/bitblas.py
@@ -4,7 +4,6 @@
 from typing import Optional
 
 import torch
-from packaging import version
 
 from vllm.logger import init_logger
 from vllm.model_executor.layers.quantization.base_config import (
@@ -111,8 +110,7 @@ class BitBLASLinearKernel(MPLinearKernel):
 
         try:
             import bitblas
-            if version.parse(bitblas.__version__) < version.parse(
-                    MINIMUM_BITBLAS_VERSION):
+            if bitblas.__version__ < MINIMUM_BITBLAS_VERSION:
                 raise ImportError(
                     "bitblas version is wrong. Please "
                     f"install bitblas>={MINIMUM_BITBLAS_VERSION}")

