diff --git a/vllm/multimodal/audio.py b/vllm/multimodal/audio.py
index d7e9d402a..fbb29276f 100644
--- a/vllm/multimodal/audio.py
+++ b/vllm/multimodal/audio.py
@@ -83,16 +83,6 @@ class AudioResampler:
 
 class AudioMediaIO(MediaIO[tuple[npt.NDArray, float]]):
 
-    def __init__(self, **kwargs) -> None:
-        super().__init__()
-
-        # `kwargs` contains custom arguments from
-        # --media-io-kwargs for this modality.
-        # They can be passed to the underlying
-        # media loaders (e.g. custom implementations)
-        # for flexible control.
-        self.kwargs = kwargs
-
     def load_bytes(self, data: bytes) -> tuple[npt.NDArray, float]:
         return librosa.load(BytesIO(data), sr=None)
 
@@ -106,7 +96,7 @@ class AudioMediaIO(MediaIO[tuple[npt.NDArray, float]]):
     def load_file(self, filepath: Path) -> tuple[npt.NDArray, float]:
         return librosa.load(filepath, sr=None)
 
-    def encode_base64(self, media: tuple[npt.NDArray, int]) -> str:
+    def encode_base64(self, media: tuple[npt.NDArray, float]) -> str:
         audio, sr = media
 
         with BytesIO() as buffer:

