diff --git a/vllm/multimodal/parse.py b/vllm/multimodal/parse.py
index 493dd3560..37f561274 100644
--- a/vllm/multimodal/parse.py
+++ b/vllm/multimodal/parse.py
@@ -16,7 +16,7 @@ from vllm.utils import LazyLoader, is_list_of
 from .audio import AudioResampler
 from .inputs import (AudioItem, HfAudioItem, HfImageItem, HfVideoItem,
                      ImageItem, ModalityData, MultiModalDataDict,
-                     MultiModalFieldConfig, MultiModalKwargsItems, VideoItem)
+                     MultiModalFieldConfig, MultiModalKwargs, VideoItem)
 
 _T = TypeVar("_T")
 _I = TypeVar("_I")
@@ -36,7 +36,7 @@ class ModalityDataItems(ABC, Generic[_T, _I]):
     def __init__(self, data: _T, modality: str) -> None:
         super().__init__()
 
-        self.data: _T = data
+        self.data = data
         self.modality = modality
 
     def __repr__(self) -> str:
@@ -157,16 +157,19 @@ class DictEmbeddingItems(ModalityDataItems[Mapping[str, torch.Tensor],
         self.fields_config = fields_config
         self.required_fields = required_fields
 
-        self._kwargs = MultiModalKwargsItems.from_hf_inputs(
+        self._kwargs = MultiModalKwargs.from_hf_inputs(
             BatchFeature(dict(data)),
             fields_config,
         )
 
     def get_count(self) -> int:
-        return len(self._kwargs[self.modality])
+        return self._kwargs.get_item_count(self.modality)
 
     def get(self, index: int) -> Mapping[str, torch.Tensor]:
-        return self._kwargs[self.modality][index].get_data()
+        return {
+            k: v.data
+            for k, v in self._kwargs.get_item(self.modality, index).items()
+        }
 
     def get_processor_data(self) -> Mapping[str, object]:
         return {}
@@ -177,9 +180,7 @@ class DictEmbeddingItems(ModalityDataItems[Mapping[str, torch.Tensor],
 
 class AudioProcessorItems(ProcessorBatchItems[HfAudioItem]):
 
-    def __init__(self, data: Optional[Sequence[HfAudioItem]]) -> None:
-        if data is None:
-            data = [None]
+    def __init__(self, data: Sequence[HfAudioItem]) -> None:
         super().__init__(data, "audio")
 
     def get_audio_length(self, item_idx: int) -> int:
@@ -200,9 +201,7 @@ class ImageSize(NamedTuple):
 
 class ImageProcessorItems(ProcessorBatchItems[HfImageItem]):
 
-    def __init__(self, data: Optional[Sequence[HfImageItem]]) -> None:
-        if data is None:
-            data = [None]
+    def __init__(self, data: Sequence[HfImageItem]) -> None:
         super().__init__(data, "image")
 
     def get_image_size(self, item_idx: int) -> ImageSize:
@@ -227,12 +226,10 @@ class VideoProcessorItems(ProcessorBatchItems[HfVideoItem]):
 
     def __init__(
         self,
-        data: Optional[Sequence[HfVideoItem]],
+        data: Sequence[HfVideoItem],
         metadata: Optional[Union[dict[str, Any],
                                  list[Optional[dict[str, Any]]]]] = None,
     ) -> None:
-        if data is None:
-            data = [None]
         super().__init__(data, "video")
         self.metadata = metadata
 
@@ -391,9 +388,6 @@ class MultiModalDataParser:
         self,
         data: ModalityData[AudioItem],
     ) -> Optional[ModalityDataItems[Any, Any]]:
-        if data is None:
-            return AudioProcessorItems(None)
-
         # also check single audio item with sampling rate
         if self._is_empty(data) or (isinstance(data, tuple)
                                     and self._is_empty(data[0])):
@@ -429,9 +423,6 @@ class MultiModalDataParser:
         self,
         data: ModalityData[ImageItem],
     ) -> Optional[ModalityDataItems[Any, Any]]:
-        if data is None:
-            return ImageProcessorItems(None)
-
         if self._is_empty(data):
             return None
 
@@ -453,9 +444,6 @@ class MultiModalDataParser:
         self,
         data: ModalityData[VideoItem],
     ) -> Optional[ModalityDataItems[Any, Any]]:
-        if data is None:
-            return VideoProcessorItems(None)
-
         if self._is_empty(data):
             return None
 

