diff --git a/megatron/core/extensions/transformer_engine.py b/megatron/core/extensions/transformer_engine.py
index d572f055..694f7e05 100644
--- a/megatron/core/extensions/transformer_engine.py
+++ b/megatron/core/extensions/transformer_engine.py
@@ -1128,12 +1128,13 @@ if HAVE_TE and is_te_min_version("1.9.0.dev0"):
 
             self._register_load_state_dict_pre_hook(merge_extra_states, with_module=True)
 
-        def forward(self, x, m_splits):
+        def forward(self, x, m_splits, fine_grained_offload: bool = False):
             """Forward."""
             _is_first_microbatch = (
                 None if self.disable_parameter_transpose_cache else self.is_first_microbatch
             )
-            out = super().forward(x, m_splits, is_first_microbatch=_is_first_microbatch)
+            out = super().forward(x, m_splits, is_first_microbatch=_is_first_microbatch, 
+                                  fine_grained_offload=fine_grained_offload)
             self.is_first_microbatch = False
 
             # TE only returns a tuple when return_bias is True, otherwise

