diff --git a/megatron/core/tokenizers/text/libraries/__init__.py b/megatron/core/tokenizers/text/libraries/__init__.py
new file mode 100755
index 00000000..e9141b81
--- /dev/null
+++ b/megatron/core/tokenizers/text/libraries/__init__.py
@@ -0,0 +1,8 @@
+# Copyright (c) 2025, NVIDIA CORPORATION. All rights reserved.
+
+from megatron.core.tokenizers.text.libraries.bytelevel_tokenizer import ByteLevelTokenizer
+from megatron.core.tokenizers.text.libraries.huggingface_tokenizer import HuggingFaceTokenizer
+from megatron.core.tokenizers.text.libraries.megatron_hf_tokenizer import MegatronHFTokenizer
+from megatron.core.tokenizers.text.libraries.null_tokenizer import NullTokenizer
+from megatron.core.tokenizers.text.libraries.sentencepiece_tokenizer import SentencePieceTokenizer
+from megatron.core.tokenizers.text.libraries.tiktoken_tokenizer import TikTokenTokenizer

