diff --git a/vllm/transformers_utils/configs/olmo3.py b/vllm/transformers_utils/configs/minimax_text_01.py
similarity index 58%
rename from vllm/transformers_utils/configs/olmo3.py
rename to vllm/transformers_utils/configs/minimax_text_01.py
index 874507db4..e3b63dfa0 100644
--- a/vllm/transformers_utils/configs/olmo3.py
+++ b/vllm/transformers_utils/configs/minimax_text_01.py
@@ -1,60 +1,48 @@
 # SPDX-License-Identifier: Apache-2.0
 # SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+""" MiniMaxText01 model configuration"""
 
 from transformers.configuration_utils import PretrainedConfig
 
 
-class Olmo3Config(PretrainedConfig):
-
-    model_type = "olmo3"
+class MiniMaxText01Config(PretrainedConfig):
+    model_type = "MiniMaxText01"
     keys_to_ignore_at_inference = ["past_key_values"]
 
     def __init__(
         self,
-        vocab_size=50304,
+        vocab_size=32000,
         hidden_size=4096,
-        intermediate_size=11008,
+        intermediate_size=14336,
         num_hidden_layers=32,
         num_attention_heads=32,
-        num_key_value_heads=None,
+        num_key_value_heads=8,
         hidden_act="silu",
-        max_position_embeddings=2048,
+        max_position_embeddings=4096 * 32,
         initializer_range=0.02,
+        rms_norm_eps=1e-5,
         use_cache=True,
-        pad_token_id=1,
+        pad_token_id=None,
         bos_token_id=None,
-        eos_token_id=50279,
+        eos_token_id=None,
         tie_word_embeddings=False,
-        rope_theta=10000.0,
-        rope_scaling=None,
-        attention_bias=False,
+        rope_theta=1e6,
+        sliding_window=None,
         attention_dropout=0.0,
-        rms_norm_eps=1e-5,
-        sliding_window=4096,
-        layer_types=None,
+        num_experts_per_tok=2,
+        num_local_experts=8,
+        output_router_logits=False,
+        router_aux_loss_coef=0.001,
+        router_jitter_noise=0.0,
         **kwargs,
     ):
-        # This model uses Olmo3ForCausalLM in transformers but Olmo2ForCausalLM
-        # in vLLM.
-        if "architectures" not in kwargs:
-            kwargs["architectures"] = ["Olmo2ForCausalLM"]
-        elif "Olmo3ForCausalLM" in kwargs["architectures"]:
-            kwargs["architectures"].remove("Olmo3ForCausalLM")
-            kwargs["architectures"].append("Olmo2ForCausalLM")
-
-        super().__init__(
-            pad_token_id=pad_token_id,
-            bos_token_id=bos_token_id,
-            eos_token_id=eos_token_id,
-            tie_word_embeddings=tie_word_embeddings,
-            **kwargs,
-        )
         self.vocab_size = vocab_size
         self.max_position_embeddings = max_position_embeddings
         self.hidden_size = hidden_size
         self.intermediate_size = intermediate_size
         self.num_hidden_layers = num_hidden_layers
         self.num_attention_heads = num_attention_heads
+        self.sliding_window = sliding_window
 
         # for backward compatibility
         if num_key_value_heads is None:
@@ -63,18 +51,20 @@ class Olmo3Config(PretrainedConfig):
         self.num_key_value_heads = num_key_value_heads
         self.hidden_act = hidden_act
         self.initializer_range = initializer_range
+        self.rms_norm_eps = rms_norm_eps
         self.use_cache = use_cache
         self.rope_theta = rope_theta
-        self.rope_scaling = rope_scaling
-        self.attention_bias = attention_bias
         self.attention_dropout = attention_dropout
 
-        self.rms_norm_eps = rms_norm_eps
-
-        self.sliding_window = sliding_window
-        self.layer_types = layer_types
-        if self.layer_types is None:
-            self.layer_types = [
-                "sliding_attention" if (i + 1) % 4 != 0 else "full_attention"
-                for i in range(self.num_hidden_layers)
-            ]
+        self.num_experts_per_tok = num_experts_per_tok
+        self.num_local_experts = num_local_experts
+        self.output_router_logits = output_router_logits
+        self.router_aux_loss_coef = router_aux_loss_coef
+        self.router_jitter_noise = router_jitter_noise
+        super().__init__(
+            pad_token_id=pad_token_id,
+            bos_token_id=bos_token_id,
+            eos_token_id=eos_token_id,
+            tie_word_embeddings=tie_word_embeddings,
+            **kwargs,
+        )

