diff --git a/vllm/v1/sample/metadata.py b/vllm/v1/sample/metadata.py
index 9d6a87cea..ab13b288a 100644
--- a/vllm/v1/sample/metadata.py
+++ b/vllm/v1/sample/metadata.py
@@ -6,8 +6,6 @@ from typing import Optional
 
 import torch
 
-from vllm.v1.sample.logits_processor import LogitsProcessors
-
 
 @dataclass
 class SamplingMetadata:
@@ -18,6 +16,7 @@ class SamplingMetadata:
 
     top_p: Optional[torch.Tensor]
     top_k: Optional[torch.Tensor]
+    min_p: Optional[torch.Tensor]
 
     generators: dict[int, torch.Generator]
 
@@ -32,12 +31,14 @@ class SamplingMetadata:
 
     output_token_ids: list[list[int]]
 
+    # req_index -> (min_tokens, stop_token_ids)
+    min_tokens: dict[int, tuple[int, set[int]]]
+
+    logit_bias: list[Optional[dict[int, float]]]
+
     # `allowed_token_ids_mask` is a 2D bool tensor of shape (max batch size,
     # vocab size).
     allowed_token_ids_mask: Optional[torch.Tensor]
 
     # req_index -> bad_words_token_ids
     bad_words_token_ids: dict[int, list[list[int]]]
-
-    # Loaded logits processors
-    logitsprocs: LogitsProcessors

