diff --git a/megatron/training/yaml_arguments.py b/megatron/training/yaml_arguments.py
index 390d503ee..d53df2210 100644
--- a/megatron/training/yaml_arguments.py
+++ b/megatron/training/yaml_arguments.py
@@ -410,7 +410,9 @@ def core_transformer_config_from_yaml(args, transfomer_key = "language_model"):
     kw_args['deallocate_pipeline_outputs'] = True
     kw_args['pipeline_dtype'] = kw_args['params_dtype']
     kw_args['batch_p2p_comm'] = not args.overlap_p2p_comm 
-    
+   
+    kw_args['share_embeddings_and_output_weights'] = not args.untie_embeddings_and_output_weights
+
     assert args.activation_func in ["swiglu","squaredrelu","gelu"], f"{args.activation_func} is not a supported activation function"
     if args.activation_func == "swiglu":
         kw_args['activation_func'] = F.silu
