diff --git a/flagscale/runner/elastic/monitor_launcher.py b/flagscale/runner/elastic/monitor_launcher.py
index 67049ff0..7954f477 100644
--- a/flagscale/runner/elastic/monitor_launcher.py
+++ b/flagscale/runner/elastic/monitor_launcher.py
@@ -10,6 +10,8 @@ import subprocess
 import sys
 import time
 
+from pathlib import Path
+
 from omegaconf import OmegaConf
 
 from flagscale.runner.elastic.monitor_service import MonitorService
@@ -52,8 +54,6 @@ def main():
     parser.add_argument(
         "--pid-file", required=True, help="The path of the PID file for the training process"
     )
-    parser.add_argument("--host", required=True, help="Hostname or IP of this node")
-    parser.add_argument("--node-rank", type=int, required=True, help="Node rank of this node")
     parser.add_argument(
         "--no-shared-fs", action="store_true", help="Whether it is in non-shared file system mode"
     )
@@ -90,12 +90,12 @@ def main():
 
     # Create dummy runners and monitoring services
     runner = MonitorRunner(config, args.pid_file)
-    monitor = MonitorService(
-        config, runner, interval=args.interval, host=args.host, node_rank=args.node_rank
-    )
+    monitor = MonitorService(config, runner, interval=args.interval)
 
     # Start the monitoring service
-    monitor.start_monitoring()
+    monitor.start_monitoring(
+        enable_log_collection=args.enable_log_collection, enable_diagnostic=args.enable_diagnostic
+    )
 
     logger.info(f"The monitoring service has been started. Interval: {args.interval} seconds")
     logger.info(f"PID file: {args.pid_file}")

