diff --git a/vllm/distributed/device_communicators/pynccl.py b/vllm/distributed/device_communicators/pynccl.py
index 76fe9a932..294862929 100644
--- a/vllm/distributed/device_communicators/pynccl.py
+++ b/vllm/distributed/device_communicators/pynccl.py
@@ -17,38 +17,6 @@ from vllm.utils import current_stream
 
 logger = init_logger(__name__)
 
-_NCCL_SYMM_OPS_REGISTERED = False
-
-
-def register_nccl_symmetric_ops(pynccl_comm):
-    from vllm.distributed.device_communicators.pynccl_allocator import (
-        nccl_symm_mem_context)
-    from vllm.utils import direct_register_custom_op
-
-    global _NCCL_SYMM_OPS_REGISTERED
-    if _NCCL_SYMM_OPS_REGISTERED:
-        return
-    _NCCL_SYMM_OPS_REGISTERED = True
-
-    def all_reduce_symmetric_with_copy_impl(
-            input_tensor: torch.Tensor) -> torch.Tensor:
-        with nccl_symm_mem_context(pynccl_comm):
-            symm_input = torch.empty_like(input_tensor)
-            symm_output = torch.empty_like(input_tensor)
-        symm_input.copy_(input_tensor)
-        symm_output = pynccl_comm.all_reduce(symm_input, symm_output)
-        return symm_output
-
-    def all_reduce_symmetric_with_copy_fake(
-            input_tensor: torch.Tensor) -> torch.Tensor:
-        return torch.empty_like(input_tensor)
-
-    direct_register_custom_op(
-        op_name="all_reduce_symmetric_with_copy",
-        op_func=all_reduce_symmetric_with_copy_impl,
-        fake_impl=all_reduce_symmetric_with_copy_fake,
-    )
-
 
 class PyNcclCommunicator:
 
@@ -63,7 +31,7 @@ class PyNcclCommunicator:
             group: the process group to work on. If None, it will use the
                 default process group.
             device: the device to bind the PyNcclCommunicator to. If None,
-                it will be bound to f"cuda:{local_rank}".
+                it will be bind to f"cuda:{local_rank}".
             library_path: the path to the NCCL library. If None, it will
                 use the default library path.
         It is the caller's responsibility to make sure each communicator
@@ -99,7 +67,6 @@ class PyNcclCommunicator:
         self.available = True
         self.disabled = False
 
-        self.nccl_version = self.nccl.ncclGetRawVersion()
         logger.info("vLLM is using nccl==%s", self.nccl.ncclGetVersion())
 
         if self.rank == 0:
@@ -142,7 +109,6 @@ class PyNcclCommunicator:
 
     def all_reduce(self,
                    in_tensor: torch.Tensor,
-                   out_tensor: torch.Tensor = None,
                    op: ReduceOp = ReduceOp.SUM,
                    stream=None) -> torch.Tensor:
         if self.disabled:
@@ -154,8 +120,7 @@ class PyNcclCommunicator:
             f"this nccl communicator is created to work on {self.device}, "
             f"but the input tensor is on {in_tensor.device}")
 
-        if out_tensor is None:
-            out_tensor = torch.empty_like(in_tensor)
+        out_tensor = torch.empty_like(in_tensor)
 
         if stream is None:
             stream = current_stream()
@@ -187,40 +152,6 @@ class PyNcclCommunicator:
             ncclDataTypeEnum.from_torch(input_tensor.dtype), self.comm,
             cudaStream_t(stream.cuda_stream))
 
-    def all_gatherv(
-        self,
-        output_tensor: torch.Tensor,
-        input_tensor: torch.Tensor,
-        sizes: list[int],
-        stream=None,
-    ):
-        if self.disabled:
-            return
-        # nccl communicator created on a specific device
-        # will only work on tensors on the same device
-        # otherwise it will cause "illegal memory access"
-        assert input_tensor.device == self.device, (
-            f"this nccl communicator is created to work on {self.device}, "
-            f"but the input tensor is on {input_tensor.device}")
-        if stream is None:
-            stream = current_stream()
-        assert output_tensor.shape[0] == sum(sizes)
-        split_offset = 0
-        self.nccl.ncclGroupStart()
-        for root, split_size in enumerate(sizes):
-            dst_slice = output_tensor[split_offset:split_offset + split_size]
-            self.nccl.ncclBroadcast(
-                buffer_type(input_tensor.data_ptr()),
-                buffer_type(dst_slice.data_ptr()),
-                dst_slice.numel(),
-                ncclDataTypeEnum.from_torch(input_tensor.dtype),
-                root,
-                self.comm,
-                cudaStream_t(stream.cuda_stream),
-            )
-            split_offset += split_size
-        self.nccl.ncclGroupEnd()
-
     def reduce_scatter(self,
                        output_tensor: torch.Tensor,
                        input_tensor: torch.Tensor,
@@ -243,38 +174,6 @@ class PyNcclCommunicator:
             ncclRedOpTypeEnum.from_torch(op), self.comm,
             cudaStream_t(stream.cuda_stream))
 
-    def reduce_scatterv(
-        self,
-        output_tensor: torch.Tensor,
-        input_tensor: torch.Tensor,
-        sizes: list[int],
-        op: ReduceOp = ReduceOp.SUM,
-        stream=None,
-    ):
-        if self.disabled:
-            return
-        # nccl communicator created on a specific device
-        # will only work on tensors on the same device
-        # otherwise it will cause "illegal memory access"
-        assert input_tensor.device == self.device, (
-            f"this nccl communicator is created to work on {self.device}, "
-            f"but the input tensor is on {input_tensor.device}")
-        if stream is None:
-            stream = current_stream()
-
-        split_offset = 0
-        self.nccl.ncclGroupStart()
-        for root, split_size in enumerate(sizes):
-            chunk = input_tensor[split_offset:split_offset + split_size, ...]
-            self.nccl.ncclReduce(
-                buffer_type(chunk.data_ptr()),
-                buffer_type(output_tensor.data_ptr()), chunk.numel(),
-                ncclDataTypeEnum.from_torch(input_tensor.dtype),
-                ncclRedOpTypeEnum.from_torch(op), root, self.comm,
-                cudaStream_t(stream.cuda_stream))
-            split_offset += split_size
-        self.nccl.ncclGroupEnd()
-
     def send(self, tensor: torch.Tensor, dst: int, stream=None):
         if self.disabled:
             return
@@ -317,24 +216,3 @@ class PyNcclCommunicator:
         self.nccl.ncclBroadcast(sendbuff, recvbuff, tensor.numel(),
                                 ncclDataTypeEnum.from_torch(tensor.dtype), src,
                                 self.comm, cudaStream_t(stream.cuda_stream))
-
-    def group_start(self):
-        self.nccl.ncclGroupStart()
-
-    def group_end(self):
-        self.nccl.ncclGroupEnd()
-
-    def register_comm_window(self, tensor: torch.Tensor):
-        return self.nccl.ncclCommWindowRegister(
-            self.comm,
-            buffer_type(tensor.data_ptr()),
-            tensor.numel() * tensor.element_size(),
-            1,
-        )
-
-    def register_comm_window_raw(self, ptr: int, size: int):
-        return self.nccl.ncclCommWindowRegister(self.comm, buffer_type(ptr),
-                                                size, 1)
-
-    def deregister_comm_window(self, window):
-        return self.nccl.ncclCommWindowDeregister(self.comm, window)

