diff --git a/vllm/logits_process.py b/vllm/logits_process.py
index 48f7e7495..5967d0836 100644
--- a/vllm/logits_process.py
+++ b/vllm/logits_process.py
@@ -1,6 +1,6 @@
 # SPDX-License-Identifier: Apache-2.0
 # SPDX-FileCopyrightText: Copyright contributors to the vLLM project
-from collections.abc import Sequence
+
 from typing import Callable, Union
 
 import torch
@@ -55,7 +55,7 @@ class NoBadWordsLogitsProcessor:
 
     def __call__(
         self,
-        past_tokens_ids: Sequence[int],
+        past_tokens_ids: Union[list[int], tuple[int]],
         logits: torch.FloatTensor,
     ) -> torch.Tensor:
         if self.word_bias is None:

