diff --git a/vllm/tracing.py b/vllm/tracing.py
index 7537e9901..6a287d82b 100644
--- a/vllm/tracing.py
+++ b/vllm/tracing.py
@@ -119,11 +119,6 @@ class SpanAttributes:
     # forward, block/sync across workers, cpu-gpu sync time and sampling time.
     GEN_AI_LATENCY_TIME_IN_MODEL_EXECUTE = (
         "gen_ai.latency.time_in_model_execute")
-    GEN_AI_LATENCY_TIME_IN_MODEL_PREFILL = \
-        "gen_ai.latency.time_in_model_prefill"
-    GEN_AI_LATENCY_TIME_IN_MODEL_DECODE = "gen_ai.latency.time_in_model_decode"
-    GEN_AI_LATENCY_TIME_IN_MODEL_INFERENCE = \
-        "gen_ai.latency.time_in_model_inference"
 
 
 def contains_trace_headers(headers: Mapping[str, str]) -> bool:

