diff --git a/vllm/assets/video.py b/vllm/assets/video.py
index 5c9e403c4..16412121c 100644
--- a/vllm/assets/video.py
+++ b/vllm/assets/video.py
@@ -59,9 +59,7 @@ def video_to_ndarrays(path: str, num_frames: int = -1) -> npt.NDArray:
         if idx in frame_indices:  # only decompress needed
             ret, frame = cap.retrieve()
             if ret:
-                # OpenCV uses BGR format, we need to convert it to RGB
-                # for PIL and transformers compatibility
-                frames.append(cv2.cvtColor(frame, cv2.COLOR_BGR2RGB))
+                frames.append(frame)
 
     frames = np.stack(frames)
     if len(frames) < num_frames:
@@ -73,10 +71,13 @@ def video_to_ndarrays(path: str, num_frames: int = -1) -> npt.NDArray:
 def video_to_pil_images_list(path: str,
                              num_frames: int = -1) -> list[Image.Image]:
     frames = video_to_ndarrays(path, num_frames)
-    return [Image.fromarray(frame) for frame in frames]
+    return [
+        Image.fromarray(cv2.cvtColor(frame, cv2.COLOR_BGR2RGB))
+        for frame in frames
+    ]
 
 
-def video_get_metadata(path: str, num_frames: int = -1) -> dict[str, Any]:
+def video_get_metadata(path: str) -> dict[str, Any]:
     cap = cv2.VideoCapture(path)
     if not cap.isOpened():
         raise ValueError(f"Could not open video file {path}")
@@ -85,18 +86,11 @@ def video_get_metadata(path: str, num_frames: int = -1) -> dict[str, Any]:
     fps = cap.get(cv2.CAP_PROP_FPS)
     duration = total_frames / fps if fps > 0 else 0
 
-    if num_frames == -1 or num_frames > total_frames:
-        num_frames = total_frames
-
     metadata = {
-        "total_num_frames": num_frames,
+        "total_num_frames": total_frames,
         "fps": fps,
         "duration": duration,
-        "video_backend": "opencv",
-        "frames_indices": list(range(num_frames)),
-        # extra field used to control hf processor's video
-        # sampling behavior
-        "do_sample_frames": num_frames == total_frames,
+        "video_backend": "opencv"
     }
     return metadata
 
@@ -117,23 +111,22 @@ class VideoAsset:
     def filename(self) -> str:
         return self._NAME_TO_FILE[self.name]
 
-    @property
-    def video_path(self) -> str:
-        return download_video_asset(self.filename)
-
     @property
     def pil_images(self) -> list[Image.Image]:
-        ret = video_to_pil_images_list(self.video_path, self.num_frames)
+        video_path = download_video_asset(self.filename)
+        ret = video_to_pil_images_list(video_path, self.num_frames)
         return ret
 
     @property
     def np_ndarrays(self) -> npt.NDArray:
-        ret = video_to_ndarrays(self.video_path, self.num_frames)
+        video_path = download_video_asset(self.filename)
+        ret = video_to_ndarrays(video_path, self.num_frames)
         return ret
 
     @property
     def metadata(self) -> dict[str, Any]:
-        ret = video_get_metadata(self.video_path, self.num_frames)
+        video_path = download_video_asset(self.filename)
+        ret = video_get_metadata(video_path)
         return ret
 
     def get_audio(self, sampling_rate: Optional[float] = None) -> npt.NDArray:
@@ -142,4 +135,5 @@ class VideoAsset:
         
         See also: examples/offline_inference/qwen2_5_omni/only_thinker.py
         """
-        return librosa.load(self.video_path, sr=sampling_rate)[0]
+        video_path = download_video_asset(self.filename)
+        return librosa.load(video_path, sr=sampling_rate)[0]

