diff --git a/vllm/attention/ops/paged_attn.py b/vllm/attention/ops/paged_attn.py
index 539b57e41..c6d1501e2 100644
--- a/vllm/attention/ops/paged_attn.py
+++ b/vllm/attention/ops/paged_attn.py
@@ -6,14 +6,9 @@ from typing import List, Optional, Tuple
 
 import torch
 
-from vllm.platforms import current_platform
+from vllm import _custom_ops as ops
 from vllm.triton_utils import HAS_TRITON
 
-if current_platform.is_cuda_alike():
-    from vllm import _custom_ops as ops
-elif current_platform.is_xpu():
-    from vllm._ipex_ops import ipex_ops as ops
-
 if HAS_TRITON:
     from vllm.attention.ops.prefix_prefill import context_attention_fwd
 
@@ -50,7 +45,6 @@ class PagedAttention:
         block_size: int,
         num_kv_heads: int,
         head_size: int,
-        cache_dtype_str: str = "auto",
     ) -> Tuple[int, ...]:
         return (2, num_blocks, block_size * num_kv_heads * head_size)
 

