diff --git a/vllm/v1/sample/tpu/metadata.py b/vllm/v1/sample/tpu/metadata.py
index 6491c84f6..4c1ac4895 100644
--- a/vllm/v1/sample/tpu/metadata.py
+++ b/vllm/v1/sample/tpu/metadata.py
@@ -5,7 +5,7 @@ from typing import Optional
 
 import torch
 
-from vllm.v1.worker.tpu_input_batch import InputBatch
+from vllm.v1.worker.gpu_input_batch import InputBatch
 
 DEFAULT_SAMPLING_PARAMS = dict(
     temperature=-1.0,

