diff --git a/vllm/lora/request.py b/vllm/lora/request.py
index 523525d46..5bbba7830 100644
--- a/vllm/lora/request.py
+++ b/vllm/lora/request.py
@@ -6,6 +6,8 @@ from typing import Optional
 
 import msgspec
 
+from vllm.adapter_commons.request import AdapterRequest
+
 
 class LoRARequest(
         msgspec.Struct,
@@ -22,6 +24,8 @@ class LoRARequest(
     lora_int_id must be globally unique for a given adapter.
     This is currently not enforced in vLLM.
     """
+    __metaclass__ = AdapterRequest
+
     lora_name: str
     lora_int_id: int
     lora_path: str = ""
@@ -31,8 +35,6 @@ class LoRARequest(
     tensorizer_config_dict: Optional[dict] = None
 
     def __post_init__(self):
-        if self.lora_int_id < 1:
-            raise ValueError(f"id must be > 0, got {self.lora_int_id}")
         if self.lora_local_path:
             warnings.warn(
                 "The 'lora_local_path' attribute is deprecated "

