diff --git a/vllm/lora/utils.py b/vllm/lora/utils.py
index 10ba390bf..ee196e3f6 100644
--- a/vllm/lora/utils.py
+++ b/vllm/lora/utils.py
@@ -2,7 +2,7 @@
 # SPDX-FileCopyrightText: Copyright contributors to the vLLM project
 
 import os
-from typing import TYPE_CHECKING, Optional, Union
+from typing import Optional, Union
 
 import huggingface_hub
 import regex as re
@@ -11,33 +11,30 @@ from huggingface_hub.utils import (EntryNotFoundError, HfHubHTTPError,
 from torch import nn
 from transformers import PretrainedConfig
 
-from vllm.config.lora import LoRAConfig
+from vllm.config import LoRAConfig
 from vllm.logger import init_logger
+from vllm.lora.fully_sharded_layers import (
+    ColumnParallelLinearWithShardedLoRA,
+    MergedColumnParallelLinearWithShardedLoRA,
+    MergedQKVParallelLinearWithShardedLoRA, QKVParallelLinearWithShardedLoRA,
+    RowParallelLinearWithShardedLoRA)
 # being imported for _all_lora_classes below
 # yapf conflicts with isort for this block
 # yapf: disable
 from vllm.lora.layers import (BaseLayerWithLoRA, ColumnParallelLinearWithLoRA,
-                              ColumnParallelLinearWithShardedLoRA,
+                              LinearScalingRotaryEmbeddingWithLoRA,
                               LogitsProcessorWithLoRA,
                               MergedColumnParallelLinearWithLoRA,
-                              MergedColumnParallelLinearWithShardedLoRA,
                               MergedQKVParallelLinearWithLoRA,
-                              MergedQKVParallelLinearWithShardedLoRA,
                               QKVParallelLinearWithLoRA,
-                              QKVParallelLinearWithShardedLoRA,
                               ReplicatedLinearWithLoRA,
                               RowParallelLinearWithLoRA,
-                              RowParallelLinearWithShardedLoRA,
                               VocabParallelEmbeddingWithLoRA)
 from vllm.model_executor.layers.linear import LinearBase
-
 # yapf: enable
-
-if TYPE_CHECKING:
-    from vllm.model_executor.layers.logits_processor import LogitsProcessor
-    from vllm.model_executor.layers.vocab_parallel_embedding import (
-        ParallelLMHead)
-    from vllm.model_executor.models.utils import WeightsMapper
+from vllm.model_executor.layers.logits_processor import LogitsProcessor
+from vllm.model_executor.layers.vocab_parallel_embedding import ParallelLMHead
+from vllm.model_executor.models.utils import WeightsMapper
 
 logger = init_logger(__name__)
 
@@ -55,6 +52,7 @@ _all_lora_classes: set[type[BaseLayerWithLoRA]] = {
     MergedColumnParallelLinearWithShardedLoRA,
     MergedQKVParallelLinearWithShardedLoRA,
     RowParallelLinearWithShardedLoRA,
+    LinearScalingRotaryEmbeddingWithLoRA,
 }
 
 
@@ -77,8 +75,8 @@ def from_layer(layer: nn.Module,
 
 
 def from_layer_logits_processor(
-    layer: "LogitsProcessor",
-    lm_head: "ParallelLMHead",
+    layer: LogitsProcessor,
+    lm_head: ParallelLMHead,
     max_loras: int,
     lora_config: LoRAConfig,
     model_config: Optional[PretrainedConfig] = None,
@@ -100,8 +98,8 @@ def replace_submodule(model: nn.Module, module_name: str,
 
 
 def parse_fine_tuned_lora_name(
-    name: str,
-    weights_mapper: Optional["WeightsMapper"] = None
+        name: str,
+        weights_mapper: Optional[WeightsMapper] = None
 ) -> tuple[str, bool, bool]:
     """Parse the name of lora weights.
 
@@ -186,20 +184,16 @@ def get_supported_lora_modules(model: nn.Module) -> list[str]:
     """
     In vLLM, all linear layers support LoRA.
     """
-
     supported_lora_modules: set[str] = set()
+    # step1: traverse the model to get all the linear subfixes.
     for name, module in model.named_modules():
-        # get the embedding modules if the module's embedding_modules
-        # is not empty.
-        embedding_modules = getattr(module, "embedding_modules", None)
-        if embedding_modules is not None:
-            for name in embedding_modules:
-                supported_lora_modules.add(name)
-
-        # get all the linear subfixes.
         if isinstance(module, (LinearBase, )):
             supported_lora_modules.add(name.split(".")[-1])
-
+    # step 2: get the embedding modules if the model's mbedding_modules
+    # is not empty.
+    if model.embedding_modules:
+        for name in model.embedding_modules:
+            supported_lora_modules.add(name)
     return list(supported_lora_modules)
 
 
@@ -239,7 +233,7 @@ def get_adapter_absolute_path(lora_path: str) -> str:
     except (HfHubHTTPError, RepositoryNotFoundError, EntryNotFoundError,
             HFValidationError):
         # Handle errors that may occur during the download
-        # Return original path instead of throwing error here
+        # Return original path instead instead of throwing error here
         logger.exception("Error downloading the HuggingFace model")
         return lora_path
 

