diff --git a/tools/patch/patch.py b/tools/patch/patch.py
index 8199762b..32cb9737 100644
--- a/tools/patch/patch.py
+++ b/tools/patch/patch.py
@@ -484,22 +484,6 @@ def validate_patch_args(device_type, task, commit, main_path):
         ), "The args commit, device_type, task must not be None."
 
 
-def domain_to_backends(domain) -> str:
-    """
-    Convert domain to backends
-    Args:
-        domain (str): Domain name provided by the user.
-    Returns:
-        str: Backend names separated by commas.
-    """
-    input_lower = domain.lower()
-
-    if input_lower in ["robo", "robotics"]:
-        return ",".join(["Megatron-LM", "Megatron-Energon"])
-    else:
-        raise ValueError(f'Unsupported domain {domain}')
-
-
 def normalize_backend(backend):
     """
     Normalize backend to standard backend names
@@ -529,6 +513,8 @@ def normalize_backend(backend):
         return "omniinfer"
     elif input_lower in ["verl"]:
         return "verl"
+    elif input_lower in ["lerobot", "Lerobot"]:
+        return "lerobot"
 
     raise ValueError(f'Unsupported backend {backend}')
 

