diff --git a/vllm/lora/ops/triton_ops/lora_expand_op.py b/vllm/lora/ops/triton_ops/lora_expand_op.py
index 467cbaa8a..9e1f90e75 100644
--- a/vllm/lora/ops/triton_ops/lora_expand_op.py
+++ b/vllm/lora/ops/triton_ops/lora_expand_op.py
@@ -8,10 +8,11 @@ https://arxiv.org/abs/2310.18547
 """
 
 import torch
+import triton
+import triton.language as tl
 
 from vllm.lora.ops.triton_ops.kernel_utils import do_expand_kernel
 from vllm.lora.ops.triton_ops.utils import _get_lora_b_ptr
-from vllm.triton_utils import tl, triton
 from vllm.utils import direct_register_custom_op
 
 

