diff --git a/megatron/training/initialize.py b/megatron/training/initialize.py
index b4b5aa02..fc4fcce5 100644
--- a/megatron/training/initialize.py
+++ b/megatron/training/initialize.py
@@ -55,8 +55,8 @@ def initialize_megatron(
     (optionally, only when args.lazy_mpu_init == True)
     """
     if not allow_no_cuda:
-        # Make sure cuda is available.
-        assert torch.cuda.is_available(), "Megatron requires CUDA."
+        # Make sure musa is available.
+        assert torch.musa.is_available(), "Megatron requires CUDA."
 
     # Parse arguments
     if parsed_args is None:
@@ -135,7 +135,7 @@ def initialize_megatron(
         if args.num_experts is not None:
             from megatron.core.transformer.moe.router import MoEAuxLossAutoScaler
 
-            MoEAuxLossAutoScaler.set_loss_scale(torch.ones(1, device=torch.cuda.current_device()))
+            MoEAuxLossAutoScaler.set_loss_scale(torch.ones(1, device=torch.musa.current_device()))
 
     if skip_mpu_initialization:
         return None
@@ -298,7 +298,7 @@ def _initialize_distributed(get_embedding_ranks, get_position_embedding_ranks, s
     """Initialize torch.distributed and core model parallel."""
     args = get_args()
 
-    device_count = torch.cuda.device_count()
+    device_count = torch.musa.device_count()
     if torch.distributed.is_initialized():
 
         if args.rank == 0:
@@ -315,14 +315,14 @@ def _initialize_distributed(get_embedding_ranks, get_position_embedding_ranks, s
             print("> initializing torch distributed ...", flush=True)
         # Manually set the device ids.
         if device_count > 0:
-            torch.cuda.set_device(args.local_rank)
-            device_id = torch.device(f'cuda:{args.local_rank}')
+            torch.musa.set_device(args.local_rank)
+            device_id = torch.device(f'musa:{args.local_rank}')
         else:
             device_id = None
 
-        # Set to non-default stream for cudagraph capturing.
+        # Set to non-default stream for musagraph capturing.
         if args.external_cuda_graph:
-            torch.cuda.set_stream(torch.cuda.Stream())
+            torch.musa.set_stream(torch.musa.Stream())
 
         # Call the init process
         init_process_group_kwargs = {
@@ -416,7 +416,7 @@ def _set_random_seed(
         random.seed(seed)
         np.random.seed(seed)
         torch.manual_seed(seed)
-        if torch.cuda.device_count() > 0:
+        if torch.musa.device_count() > 0:
             tensor_parallel.model_parallel_cuda_manual_seed(
                 seed, te_rng_tracker, inference_rng_tracker, use_cudagraphable_rng
             )
@@ -469,7 +469,7 @@ def _warmup_jit_function():
 
     # Warmup fused bias+gelu
     bias = torch.rand(
-        args.ffn_hidden_size // args.tensor_model_parallel_size, dtype=dtype, device="cuda"
+        args.ffn_hidden_size // args.tensor_model_parallel_size, dtype=dtype, device="musa"
     )
     input = torch.rand(
         (
@@ -478,7 +478,7 @@ def _warmup_jit_function():
             args.ffn_hidden_size // args.tensor_model_parallel_size,
         ),
         dtype=dtype,
-        device="cuda",
+        device="musa",
     )
     # Warmup JIT fusions with the input grad_enable state of both forward
     # prop and recomputation
@@ -499,14 +499,14 @@ def _warmup_jit_function():
     input = torch.rand(
         (seq_length // args.context_parallel_size, args.micro_batch_size, args.hidden_size),
         dtype=dtype,
-        device="cuda",
+        device="musa",
     )
     residual = torch.rand(
         (seq_length // args.context_parallel_size, args.micro_batch_size, args.hidden_size),
         dtype=dtype,
-        device="cuda",
+        device="musa",
     )
-    bias = torch.rand((args.hidden_size), dtype=dtype, device="cuda").expand_as(residual)
+    bias = torch.rand((args.hidden_size), dtype=dtype, device="musa").expand_as(residual)
     dropout_rate = 0.1
     # Warmup JIT fusions with the input grad_enable state of both forward
     # prop and recomputation
@@ -517,7 +517,7 @@ def _warmup_jit_function():
         for _ in range(5):
             output = bias_dropout_add_fused_train([input, bias], residual, dropout_rate)
     del bias, input, residual, output
-    torch.cuda.empty_cache()
+    torch.musa.empty_cache()
 
 
 def setup_logging() -> None:

