diff --git a/requirements/test.in b/requirements/test.in
index 451bd7387..bb4c8ecc3 100644
--- a/requirements/test.in
+++ b/requirements/test.in
@@ -50,7 +50,7 @@ tritonclient==2.51.0
 
 numba == 0.60.0; python_version == '3.9' # v0.61 doesn't support Python 3.9. Required for N-gram speculative decoding
 numba == 0.61.2; python_version > '3.9'
-numpy
+numpy==1.26.4
 runai-model-streamer==0.11.0
 runai-model-streamer-s3==0.11.0
 fastsafetensors>=0.1.10

