diff --git a/vllm/model_executor/layers/mamba/ops/ssd_chunk_state.py b/vllm/model_executor/layers/mamba/ops/ssd_chunk_state.py
index 8ee41f2cb..58bfb661d 100644
--- a/vllm/model_executor/layers/mamba/ops/ssd_chunk_state.py
+++ b/vllm/model_executor/layers/mamba/ops/ssd_chunk_state.py
@@ -17,6 +17,7 @@ from .mamba_ssm import softplus
 
 @triton.autotune(
     configs=[
+        triton.Config({'BLOCK_SIZE_H': 1}),
         triton.Config({'BLOCK_SIZE_H': 2}),
         triton.Config({'BLOCK_SIZE_H': 4}),
         triton.Config({'BLOCK_SIZE_H': 8}),
@@ -35,35 +36,41 @@ def _chunk_cumsum_fwd_kernel(
     dt_out_ptr,
     dA_cumsum_ptr,
     # Matrix dimension
+    batch,
     seqlen,
-    nheads: tl.constexpr,
-    chunk_size: tl.constexpr,
-    dt_min: tl.constexpr,
-    dt_max: tl.constexpr,
+    nheads,
+    chunk_size,
+    dt_min,
+    dt_max,
     # Strides
-    stride_dt_seqlen: tl.int64,
-    stride_dt_head: tl.constexpr,
-    stride_A_head: tl.constexpr,
-    stride_dt_bias_head: tl.constexpr,
-    stride_dt_out_head: tl.int64,
-    stride_dt_out_chunk: tl.int64,
-    stride_dt_out_csize: tl.constexpr,
-    stride_dA_cs_head: tl.int64,
-    stride_dA_cs_chunk: tl.int64,
-    stride_dA_cs_csize: tl.constexpr,
+    stride_dt_batch,
+    stride_dt_seqlen,
+    stride_dt_head,
+    stride_A_head,
+    stride_dt_bias_head,
+    stride_dt_out_batch,
+    stride_dt_out_chunk,
+    stride_dt_out_head,
+    stride_dt_out_csize,
+    stride_dA_cs_batch,
+    stride_dA_cs_chunk,
+    stride_dA_cs_head,
+    stride_dA_cs_csize,
     # Meta-parameters
     DT_SOFTPLUS: tl.constexpr,
     HAS_DT_BIAS: tl.constexpr,
     BLOCK_SIZE_H: tl.constexpr,
     BLOCK_SIZE_CHUNK: tl.constexpr,
 ):
+    pid_b = tl.program_id(axis=0)
+
     # if dt is long, may cause problems, so use 64 bit
     # https://github.com/triton-lang/triton/issues/1058
-    pid_c = tl.program_id(axis=0).to(tl.int64)
-    pid_h = tl.program_id(axis=1)
-    dt_ptr += pid_c * chunk_size * stride_dt_seqlen
-    dt_out_ptr += pid_c * stride_dt_out_chunk
-    dA_cumsum_ptr += pid_c * stride_dA_cs_chunk
+    pid_c = tl.program_id(axis=1).to(tl.int64)
+    pid_h = tl.program_id(axis=2)
+    dt_ptr += pid_b * stride_dt_batch + pid_c * chunk_size * stride_dt_seqlen
+    dt_out_ptr += pid_b * stride_dt_out_batch + pid_c * stride_dt_out_chunk
+    dA_cumsum_ptr += pid_b * stride_dA_cs_batch + pid_c * stride_dA_cs_chunk
 
     offs_h = pid_h * BLOCK_SIZE_H + tl.arange(0, BLOCK_SIZE_H)
     offs_c = tl.arange(0, BLOCK_SIZE_CHUNK)
@@ -87,8 +94,9 @@ def _chunk_cumsum_fwd_kernel(
         dt += dt_bias[:, None]
     if DT_SOFTPLUS:
         dt = tl.where(dt <= 20.0, softplus(dt), dt)
-
-    dt = tl.clamp(dt, dt_min, dt_max)
+    # As of Triton 2.2.0, tl.clamp is not available yet
+    # dt = tl.clamp(dt, dt_min, dt_max)
+    dt = tl.minimum(tl.maximum(dt, dt_min), dt_max)
     dt = tl.where(
         (offs_h[:, None] < nheads) & (offs_c[None, :] < chunk_size_limit), dt,
         0.0)
@@ -190,46 +198,56 @@ def _chunk_state_fwd_kernel(
     dA_cumsum_ptr,
     seq_idx_ptr,
     # Matrix dimensions
-    hdim: tl.constexpr,
-    dstate: tl.constexpr,
-    chunk_size: tl.constexpr,
+    hdim,
+    dstate,
+    chunk_size,
+    batch,
     seqlen,
-    nheads_ngroups_ratio: tl.constexpr,
+    nheads_ngroups_ratio,
     # Strides
-    stride_x_seqlen: tl.int64,
-    stride_x_head: tl.int64,
-    stride_x_hdim: tl.constexpr,
-    stride_b_seqlen: tl.int64,
-    stride_b_head: tl.int64,
-    stride_b_dstate: tl.constexpr,
-    stride_states_chunk: tl.int64,
-    stride_states_head: tl.int64,
-    stride_states_hdim: tl.int64,
-    stride_states_dstate: tl.constexpr,
-    stride_dt_head: tl.int64,
-    stride_dt_chunk: tl.int64,
-    stride_dt_csize: tl.constexpr,
-    stride_dA_cs_head: tl.int64,
-    stride_dA_cs_chunk: tl.int64,
-    stride_dA_cs_csize: tl.constexpr,
-    stride_seq_idx_seqlen: tl.constexpr,
+    stride_x_batch,
+    stride_x_seqlen,
+    stride_x_head,
+    stride_x_hdim,
+    stride_b_batch,
+    stride_b_seqlen,
+    stride_b_head,
+    stride_b_dstate,
+    stride_states_batch,
+    stride_states_chunk,
+    stride_states_head,
+    stride_states_hdim,
+    stride_states_dstate,
+    stride_dt_batch,
+    stride_dt_chunk,
+    stride_dt_head,
+    stride_dt_csize,
+    stride_dA_cs_batch,
+    stride_dA_cs_chunk,
+    stride_dA_cs_head,
+    stride_dA_cs_csize,
+    stride_seq_idx_batch,
+    stride_seq_idx_seqlen,
     # Meta-parameters
+    HAS_SEQ_IDX: tl.constexpr,
     BLOCK_SIZE_M: tl.constexpr,
     BLOCK_SIZE_N: tl.constexpr,
     BLOCK_SIZE_K: tl.constexpr,
 ):
-    pid_c = tl.program_id(axis=1).to(tl.int64)
+    pid_bc = tl.program_id(axis=1).to(tl.int64)
+    pid_c = pid_bc // batch
+    pid_b = pid_bc - pid_c * batch
     pid_h = tl.program_id(axis=2)
     num_pid_n = tl.cdiv(dstate, BLOCK_SIZE_N)
     pid_m = tl.program_id(axis=0) // num_pid_n
     pid_n = tl.program_id(axis=0) % num_pid_n
-    b_ptr += pid_c * chunk_size * stride_b_seqlen + (
+    b_ptr += pid_b * stride_b_batch + pid_c * chunk_size * stride_b_seqlen + (
         pid_h // nheads_ngroups_ratio) * stride_b_head
-    x_ptr += pid_c * chunk_size * stride_x_seqlen + pid_h * stride_x_head
-    dt_ptr += pid_c * stride_dt_chunk + pid_h * stride_dt_head
-    dA_cumsum_ptr += pid_c * stride_dA_cs_chunk + pid_h * stride_dA_cs_head
-
-    seq_idx_ptr += pid_c * chunk_size * stride_seq_idx_seqlen
+    x_ptr += pid_b * stride_x_batch + pid_c * chunk_size * stride_x_seqlen + pid_h * stride_x_head
+    dt_ptr += pid_b * stride_dt_batch + pid_c * stride_dt_chunk + pid_h * stride_dt_head
+    dA_cumsum_ptr += pid_b * stride_dA_cs_batch + pid_c * stride_dA_cs_chunk + pid_h * stride_dA_cs_head
+    if HAS_SEQ_IDX:
+        seq_idx_ptr += pid_b * stride_seq_idx_batch + pid_c * chunk_size * stride_seq_idx_seqlen
 
     offs_m = pid_m * BLOCK_SIZE_M + tl.arange(0, BLOCK_SIZE_M)
     offs_n = pid_n * BLOCK_SIZE_N + tl.arange(0, BLOCK_SIZE_N)
@@ -242,11 +260,13 @@ def _chunk_state_fwd_kernel(
     dA_cs_last = tl.load(dA_cumsum_ptr +
                          (chunk_size - 1) * stride_dA_cs_csize).to(tl.float32)
     dA_cumsum_ptrs = dA_cumsum_ptr + offs_k * stride_dA_cs_csize
+    if HAS_SEQ_IDX:
+        seq_idx_ptrs = seq_idx_ptr + offs_k * stride_seq_idx_seqlen
 
-    seq_idx_ptrs = seq_idx_ptr + offs_k * stride_seq_idx_seqlen
     chunk_size_limit = min(chunk_size, seqlen - pid_c * chunk_size)
-    seq_idx_last = tl.load(seq_idx_ptr +
-                           (chunk_size_limit - 1) * stride_seq_idx_seqlen)
+    if HAS_SEQ_IDX:
+        seq_idx_last = tl.load(seq_idx_ptr +
+                               (chunk_size_limit - 1) * stride_seq_idx_seqlen)
 
     acc = tl.zeros((BLOCK_SIZE_M, BLOCK_SIZE_N), dtype=tl.float32)
     for k in range(0, chunk_size_limit, BLOCK_SIZE_K):
@@ -261,28 +281,29 @@ def _chunk_state_fwd_kernel(
         dA_cs_k = tl.load(dA_cumsum_ptrs,
                           mask=offs_k < chunk_size_limit - k,
                           other=0.0).to(tl.float32)
-
-        seq_idx_k = tl.load(seq_idx_ptrs,
-                            mask=offs_k < chunk_size_limit - k,
-                            other=-1)
+        if HAS_SEQ_IDX:
+            seq_idx_k = tl.load(seq_idx_ptrs,
+                                mask=offs_k < chunk_size_limit - k,
+                                other=-1)
         dt_k = tl.load(dt_ptrs, mask=offs_k < chunk_size_limit - k,
                        other=0.0).to(tl.float32)
-
-        scale = tl.where(seq_idx_k == seq_idx_last,
-                         tl.exp(dA_cs_last - dA_cs_k) * dt_k, 0.0)
+        if not HAS_SEQ_IDX:
+            scale = tl.exp(dA_cs_last - dA_cs_k) * dt_k
+        else:
+            scale = tl.where(seq_idx_k == seq_idx_last,
+                             tl.exp(dA_cs_last - dA_cs_k) * dt_k, 0.0)
         b *= scale[:, None]
         b = b.to(x_ptr.dtype.element_ty)
         acc += tl.dot(x, b)
-
         x_ptrs += BLOCK_SIZE_K * stride_x_seqlen
         b_ptrs += BLOCK_SIZE_K * stride_b_seqlen
         dt_ptrs += BLOCK_SIZE_K * stride_dt_csize
         dA_cumsum_ptrs += BLOCK_SIZE_K * stride_dA_cs_csize
-        seq_idx_ptrs += BLOCK_SIZE_K * stride_seq_idx_seqlen
-
+        if HAS_SEQ_IDX:
+            seq_idx_ptrs += BLOCK_SIZE_K * stride_seq_idx_seqlen
     states = acc.to(states_ptr.dtype.element_ty)
 
-    states_ptr += pid_c * stride_states_chunk + pid_h * stride_states_head
+    states_ptr += pid_b * stride_states_batch + pid_c * stride_states_chunk + pid_h * stride_states_head
     offs_m = pid_m * BLOCK_SIZE_M + tl.arange(0, BLOCK_SIZE_M)
     offs_n = pid_n * BLOCK_SIZE_N + tl.arange(0, BLOCK_SIZE_N)
     states_ptrs = states_ptr + (offs_m[:, None] * stride_states_hdim +
@@ -380,35 +401,36 @@ def _chunk_state_varlen_kernel(
     states_ptr,
     initstates_ptr,
     # Matrix dimensions
-    hdim: tl.constexpr,
-    dstate: tl.constexpr,
-    chunk_size: tl.constexpr,
-    nheads_ngroups_ratio: tl.constexpr,
+    hdim,
+    dstate,
+    chunk_size,
+    seqlen,
+    nheads_ngroups_ratio,
     # Strides
-    stride_x_seqlen: tl.int64,
-    stride_x_head: tl.int64,
-    stride_x_hdim: tl.constexpr,
-    stride_b_seqlen: tl.int64,
-    stride_b_head: tl.int64,
-    stride_b_dstate: tl.constexpr,
-    stride_dt_head: tl.int64,
-    stride_dt_chunk: tl.int64,
-    stride_dt_csize: tl.constexpr,
-    stride_dA_cs_head: tl.int64,
-    stride_dA_cs_chunk: tl.int64,
-    stride_dA_cs_csize: tl.constexpr,
-    stride_chunk_states_chunk: tl.int64,
-    stride_chunk_states_head: tl.int64,
-    stride_chunk_states_hdim: tl.int64,
-    stride_chunk_states_dstate: tl.constexpr,
-    stride_states_batch: tl.int64,
-    stride_states_head: tl.int64,
-    stride_states_hdim: tl.int64,
-    stride_states_dstate: tl.constexpr,
-    stride_init_states_batch: tl.int64,
-    stride_init_states_head: tl.int64,
-    stride_init_states_hdim: tl.int64,
-    stride_init_states_dstate: tl.constexpr,
+    stride_x_seqlen,
+    stride_x_head,
+    stride_x_hdim,
+    stride_b_seqlen,
+    stride_b_head,
+    stride_b_dstate,
+    stride_dt_chunk,
+    stride_dt_head,
+    stride_dt_csize,
+    stride_dA_cs_chunk,
+    stride_dA_cs_head,
+    stride_dA_cs_csize,
+    stride_chunk_states_chunk,
+    stride_chunk_states_head,
+    stride_chunk_states_hdim,
+    stride_chunk_states_dstate,
+    stride_states_batch,
+    stride_states_head,
+    stride_states_hdim,
+    stride_states_dstate,
+    stride_init_states_batch,
+    stride_init_states_head,
+    stride_init_states_hdim,
+    stride_init_states_dstate,
     # Meta-parameters
     BLOCK_SIZE_M: tl.constexpr,
     BLOCK_SIZE_N: tl.constexpr,
@@ -480,7 +502,7 @@ def _chunk_state_varlen_kernel(
         dA_cumsum_ptrs += BLOCK_SIZE_K * stride_dA_cs_csize
 
     # If the sequence starts after the last chunk idx, we don't need to add the contribution from the last chunk
-    # If HAS_INITSTATES==True need to consider two possibilities
+    # If HAS_INITSTATES==True need to consider two possiblties
     # - if start_idx < pid_c * chunk_size, then we need to take the past_states_ptrs
     # - if state_idx >= pid * chunk_size, then we need to insert initstates
     if ((start_idx < pid_c * chunk_size)  # first chunk
@@ -494,7 +516,7 @@ def _chunk_state_varlen_kernel(
                 offs_n[None, :] * stride_chunk_states_dstate)
         else:
 
-            # - this seems repetitive, buts its to help the compiler
+            # - this seems repetitve, buts its to help the compiler
             if start_idx < pid_c * chunk_size:
                 past_states_ptrs = chunk_states_ptr + (
                     offs_m[:, None] * stride_chunk_states_hdim +
@@ -537,47 +559,52 @@ def _chunk_cumsum_fwd(dt,
                       dt_bias=None,
                       dt_softplus=False,
                       dt_limit=(0.0, float("inf"))):
-    seqlen, nheads = dt.shape
+    batch, seqlen, nheads = dt.shape
     assert A.shape == (nheads, )
     if dt_bias is not None:
         assert dt_bias.shape == (nheads, )
     nchunks = math.ceil(seqlen / chunk_size)
-    dt_out = torch.empty(nheads,
+    dt_out = torch.empty(batch,
+                         nheads,
                          nchunks,
                          chunk_size,
                          device=dt.device,
                          dtype=torch.float32)
-    dA_cumsum = torch.empty(nheads,
+    dA_cumsum = torch.empty(batch,
+                            nheads,
                             nchunks,
                             chunk_size,
                             device=dt.device,
                             dtype=torch.float32)
-    grid_chunk_cs = lambda META: (nchunks,
+    grid_chunk_cs = lambda META: (batch, nchunks,
                                   triton.cdiv(nheads, META['BLOCK_SIZE_H']))
     with torch.cuda.device(dt.device.index):
         _chunk_cumsum_fwd_kernel[grid_chunk_cs](
-            dt_ptr=dt,
-            A_ptr=A,
-            dt_bias_ptr=dt_bias,
-            dt_out_ptr=dt_out,
-            dA_cumsum_ptr=dA_cumsum,
-            seqlen=seqlen,
-            nheads=nheads,
-            chunk_size=chunk_size,
-            dt_min=dt_limit[0],
-            dt_max=dt_limit[1],
-            stride_dt_seqlen=dt.stride(0),
-            stride_dt_head=dt.stride(1),
-            stride_A_head=A.stride(0),
-            stride_dt_bias_head=dt_bias.stride(0)
-            if dt_bias is not None else 0,
-            stride_dt_out_head=dt_out.stride(0),
-            stride_dt_out_chunk=dt_out.stride(1),
-            stride_dt_out_csize=dt_out.stride(2),
-            stride_dA_cs_head=dA_cumsum.stride(0),
-            stride_dA_cs_chunk=dA_cumsum.stride(1),
-            stride_dA_cs_csize=dA_cumsum.stride(2),
-            DT_SOFTPLUS=dt_softplus,
+            dt,
+            A,
+            dt_bias,
+            dt_out,
+            dA_cumsum,
+            batch,
+            seqlen,
+            nheads,
+            chunk_size,
+            dt_limit[0],
+            dt_limit[1],
+            dt.stride(0),
+            dt.stride(1),
+            dt.stride(2),
+            A.stride(0),
+            dt_bias.stride(0) if dt_bias is not None else 0,
+            dt_out.stride(0),
+            dt_out.stride(2),
+            dt_out.stride(1),
+            dt_out.stride(3),
+            dA_cumsum.stride(0),
+            dA_cumsum.stride(2),
+            dA_cumsum.stride(1),
+            dA_cumsum.stride(3),
+            dt_softplus,
             HAS_DT_BIAS=dt_bias is not None,
             BLOCK_SIZE_CHUNK=triton.next_power_of_2(chunk_size),
         )
@@ -591,57 +618,63 @@ def _chunk_state_fwd(B,
                      seq_idx=None,
                      states=None,
                      states_in_fp32=True):
-    seqlen, nheads, headdim = x.shape
-    _, nchunks, chunk_size = dt.shape
-    _, ngroups, dstate = B.shape
+    batch, seqlen, nheads, headdim = x.shape
+    _, _, nchunks, chunk_size = dt.shape
+    _, _, ngroups, dstate = B.shape
     assert nheads % ngroups == 0
-    assert B.shape == (seqlen, ngroups, dstate)
-    assert dt.shape == (nheads, nchunks, chunk_size)
+    assert B.shape == (batch, seqlen, ngroups, dstate)
+    assert dt.shape == (batch, nheads, nchunks, chunk_size)
     assert dA_cumsum.shape == dt.shape
-
-    assert seq_idx is not None
-    assert seq_idx.shape == (seqlen, )
-
+    if seq_idx is not None:
+        assert seq_idx.shape == (batch, seqlen)
     if states is not None:
-        assert states.shape == (nchunks, nheads, headdim, dstate)
+        assert states.shape == (batch, nchunks, nheads, headdim, dstate)
     else:
         states_dtype = torch.float32 if states_in_fp32 else B.dtype
-        states = torch.empty((nchunks, nheads, headdim, dstate),
+        states = torch.empty((batch, nchunks, nheads, headdim, dstate),
                              device=x.device,
                              dtype=states_dtype)
-
-    grid = lambda META: (triton.cdiv(headdim, META['BLOCK_SIZE_M']) * triton.
-                         cdiv(dstate, META['BLOCK_SIZE_N']), nchunks, nheads)
+    grid = lambda META: (
+        triton.cdiv(headdim, META['BLOCK_SIZE_M']) * triton.cdiv(
+            dstate, META['BLOCK_SIZE_N']), batch * nchunks, nheads)
     with torch.cuda.device(x.device.index):
         _chunk_state_fwd_kernel[grid](
-            x_ptr=x,
-            b_ptr=B,
-            states_ptr=states,
-            dt_ptr=dt,
-            dA_cumsum_ptr=dA_cumsum,
-            seq_idx_ptr=seq_idx,
-            hdim=headdim,
-            dstate=dstate,
-            chunk_size=chunk_size,
-            seqlen=seqlen,
-            nheads_ngroups_ratio=nheads // ngroups,
-            stride_x_seqlen=x.stride(0),
-            stride_x_head=x.stride(1),
-            stride_x_hdim=x.stride(2),
-            stride_b_seqlen=B.stride(0),
-            stride_b_head=B.stride(1),
-            stride_b_dstate=B.stride(2),
-            stride_states_chunk=states.stride(0),
-            stride_states_head=states.stride(1),
-            stride_states_hdim=states.stride(2),
-            stride_states_dstate=states.stride(3),
-            stride_dt_head=dt.stride(0),
-            stride_dt_chunk=dt.stride(1),
-            stride_dt_csize=dt.stride(2),
-            stride_dA_cs_head=dA_cumsum.stride(0),
-            stride_dA_cs_chunk=dA_cumsum.stride(1),
-            stride_dA_cs_csize=dA_cumsum.stride(2),
-            stride_seq_idx_seqlen=seq_idx.stride(0),
+            x,
+            B,
+            states,
+            dt,
+            dA_cumsum,
+            seq_idx,
+            headdim,
+            dstate,
+            chunk_size,
+            batch,
+            seqlen,
+            nheads // ngroups,
+            x.stride(0),
+            x.stride(1),
+            x.stride(2),
+            x.stride(3),
+            B.stride(0),
+            B.stride(1),
+            B.stride(2),
+            B.stride(-1),
+            states.stride(0),
+            states.stride(1),
+            states.stride(2),
+            states.stride(3),
+            states.stride(4),
+            dt.stride(0),
+            dt.stride(2),
+            dt.stride(1),
+            dt.stride(3),
+            dA_cumsum.stride(0),
+            dA_cumsum.stride(2),
+            dA_cumsum.stride(1),
+            dA_cumsum.stride(3),
+            *((seq_idx.stride(0),
+               seq_idx.stride(1)) if seq_idx is not None else (0, 0)),
+            HAS_SEQ_IDX=seq_idx is not None,
         )
     return states
 
@@ -673,52 +706,46 @@ def chunk_state_varlen(B,
                          dstate,
                          dtype=chunk_states.dtype,
                          device=chunk_states.device)
-
-    initial_states_strides = ((initial_states.stride(0),
-                               initial_states.stride(1),
-                               initial_states.stride(2),
-                               initial_states.stride(3))
-                              if initial_states is not None else (0, 0, 0, 0))
-
     grid = lambda META: (triton.cdiv(headdim, META['BLOCK_SIZE_M']) * triton.
                          cdiv(dstate, META['BLOCK_SIZE_N']), batch, nheads)
     with torch.cuda.device(x.device.index):
         _chunk_state_varlen_kernel[grid](
-            x_ptr=x,
-            b_ptr=B,
-            dt_ptr=dt,
-            dA_cumsum_ptr=dA_cumsum,
-            chunk_states_ptr=chunk_states,
-            cu_seqlens_ptr=cu_seqlens,
-            states_ptr=states,
-            initstates_ptr=initial_states,
-            hdim=headdim,
-            dstate=dstate,
-            chunk_size=chunk_size,
-            nheads_ngroups_ratio=nheads // ngroups,
-            stride_x_seqlen=x.stride(0),
-            stride_x_head=x.stride(1),
-            stride_x_hdim=x.stride(2),
-            stride_b_seqlen=B.stride(0),
-            stride_b_head=B.stride(1),
-            stride_b_dstate=B.stride(2),
-            stride_dt_head=dt.stride(0),
-            stride_dt_chunk=dt.stride(1),
-            stride_dt_csize=dt.stride(2),
-            stride_dA_cs_head=dA_cumsum.stride(0),
-            stride_dA_cs_chunk=dA_cumsum.stride(1),
-            stride_dA_cs_csize=dA_cumsum.stride(2),
-            stride_chunk_states_chunk=chunk_states.stride(0),
-            stride_chunk_states_head=chunk_states.stride(1),
-            stride_chunk_states_hdim=chunk_states.stride(2),
-            stride_chunk_states_dstate=chunk_states.stride(3),
-            stride_states_batch=states.stride(0),
-            stride_states_head=states.stride(1),
-            stride_states_hdim=states.stride(2),
-            stride_states_dstate=states.stride(3),
-            stride_init_states_batch=initial_states_strides[0],
-            stride_init_states_head=initial_states_strides[1],
-            stride_init_states_hdim=initial_states_strides[2],
-            stride_init_states_dstate=initial_states_strides[3],
+            x,
+            B,
+            dt,
+            dA_cumsum,
+            chunk_states,
+            cu_seqlens,
+            states,
+            initial_states,
+            headdim,
+            dstate,
+            chunk_size,
+            total_seqlen,
+            nheads // ngroups,
+            x.stride(0),
+            x.stride(1),
+            x.stride(2),
+            B.stride(0),
+            B.stride(1),
+            B.stride(2),
+            dt.stride(1),
+            dt.stride(0),
+            dt.stride(2),
+            dA_cumsum.stride(1),
+            dA_cumsum.stride(0),
+            dA_cumsum.stride(2),
+            chunk_states.stride(0),
+            chunk_states.stride(1),
+            chunk_states.stride(2),
+            chunk_states.stride(3),
+            states.stride(0),
+            states.stride(1),
+            states.stride(2),
+            states.stride(3),
+            *((initial_states.stride(0), initial_states.stride(1),
+               initial_states.stride(2),
+               initial_states.stride(3)) if initial_states is not None else
+              (0, 0, 0, 0)),
             HAS_INITSTATES=initial_states is not None)
     return states

