diff --git a/vllm/v1/executor/abstract.py b/vllm/v1/executor/abstract.py
index 625017d52..50b9634a4 100644
--- a/vllm/v1/executor/abstract.py
+++ b/vllm/v1/executor/abstract.py
@@ -2,7 +2,7 @@
 # SPDX-FileCopyrightText: Copyright contributors to the vLLM project
 
 from concurrent.futures import Future
-from typing import Any, Callable, Optional, Union
+from typing import Callable, Union
 
 import torch
 import torch.distributed as dist
@@ -13,10 +13,8 @@ from vllm.executor.uniproc_executor import (  # noqa
     ExecutorWithExternalLauncher as ExecutorWithExternalLauncherV0)
 from vllm.executor.uniproc_executor import (  # noqa
     UniProcExecutor as UniProcExecutorV0)
-from vllm.utils import resolve_obj_by_qualname
-from vllm.v1.core.sched.output import SchedulerOutput
 from vllm.v1.kv_cache_interface import KVCacheConfig, KVCacheSpec
-from vllm.v1.outputs import DraftTokenIds, ModelRunnerOutput
+from vllm.v1.outputs import ModelRunnerOutput
 
 FailureCallback = Callable[[], None]
 
@@ -52,13 +50,6 @@ class Executor(ExecutorBase):
             # TODO: make v1 scheduling deterministic
             # to support external launcher
             executor_class = ExecutorWithExternalLauncher
-        elif isinstance(distributed_executor_backend, str):
-            executor_class = resolve_obj_by_qualname(
-                distributed_executor_backend)
-            if not issubclass(executor_class, ExecutorBase):
-                raise TypeError(
-                    "distributed_executor_backend must be a subclass of "
-                    f"ExecutorBase. Got {executor_class}.")
         else:
             raise ValueError("Unknown distributed executor backend: "
                              f"{distributed_executor_backend}")
@@ -82,34 +73,19 @@ class Executor(ExecutorBase):
         pass
 
     def determine_available_memory(self) -> list[int]:  # in bytes
-        return self.collective_rpc("determine_available_memory")
+        output = self.collective_rpc("determine_available_memory")
+        return output
 
     def get_kv_cache_specs(self) -> list[dict[str, KVCacheSpec]]:
-        return self.collective_rpc("get_kv_cache_spec")
-
-    def collective_rpc(self,
-                       method: Union[str, Callable],
-                       timeout: Optional[float] = None,
-                       args: tuple = (),
-                       kwargs: Optional[dict] = None,
-                       non_block: bool = False) -> list[Any]:
-        raise NotImplementedError
+        output = self.collective_rpc("get_kv_cache_spec")
+        return output
 
     def execute_model(
         self,
-        scheduler_output: SchedulerOutput,
-        non_block: bool = False,
+        scheduler_output,
     ) -> Union[ModelRunnerOutput, Future[ModelRunnerOutput]]:
         output = self.collective_rpc("execute_model",
-                                     args=(scheduler_output, ),
-                                     non_block=non_block)
-        return output[0]
-
-    def execute_dummy_batch(self) -> None:
-        self.collective_rpc("execute_dummy_batch")
-
-    def take_draft_token_ids(self) -> Optional[DraftTokenIds]:
-        output = self.collective_rpc("take_draft_token_ids")
+                                     args=(scheduler_output, ))
         return output[0]
 
     @property

