diff --git a/vllm/model_executor/model_loader/base_loader.py b/vllm/model_executor/model_loader/base_loader.py
index ab538a3c9..5018c7d9a 100644
--- a/vllm/model_executor/model_loader/base_loader.py
+++ b/vllm/model_executor/model_loader/base_loader.py
@@ -5,14 +5,10 @@ from abc import ABC, abstractmethod
 import torch
 import torch.nn as nn
 
-from vllm.config import ModelConfig, VllmConfig
-from vllm.config.load import LoadConfig
-from vllm.logger import init_logger
+from vllm.config import LoadConfig, ModelConfig, VllmConfig
 from vllm.model_executor.model_loader.utils import (
     initialize_model, process_weights_after_loading, set_default_torch_dtype)
 
-logger = init_logger(__name__)
-
 
 class BaseModelLoader(ABC):
     """Base class for model loaders."""
@@ -36,16 +32,11 @@ class BaseModelLoader(ABC):
                    model_config: ModelConfig) -> nn.Module:
         """Load a model with the given configurations."""
         device_config = vllm_config.device_config
-        load_config = vllm_config.load_config
-        load_device = device_config.device if load_config.device is None else \
-                      load_config.device
-        target_device = torch.device(load_device)
+        target_device = torch.device(device_config.device)
         with set_default_torch_dtype(model_config.dtype):
             with target_device:
                 model = initialize_model(vllm_config=vllm_config,
                                          model_config=model_config)
-
-            logger.debug("Loading weights on %s ...", load_device)
             # Quantization does not happen in `load_weights` but after it
             self.load_weights(model, model_config)
             process_weights_after_loading(model, model_config, target_device)

