diff --git a/vllm/engine/metrics_types.py b/vllm/engine/metrics_types.py
index 9778ab5a8..9375dc4c4 100644
--- a/vllm/engine/metrics_types.py
+++ b/vllm/engine/metrics_types.py
@@ -16,9 +16,10 @@ do this in Python code and lazily import prometheus_client.
 import time
 from abc import ABC, abstractmethod
 from dataclasses import dataclass
-from typing import List
+from typing import List, Optional
 
 from vllm.config import SupportsMetricsInfo, VllmConfig
+from vllm.spec_decode.metrics import SpecDecodeWorkerMetrics
 
 
 @dataclass
@@ -43,7 +44,7 @@ class Stats:
     num_generation_tokens_iter: int
     num_tokens_iter: int
     time_to_first_tokens_iter: List[float]
-    inter_token_latencies_iter: List[float]
+    time_per_output_tokens_iter: List[float]
     num_preemption_iter: int
 
     # Request stats (should have _requests suffix)
@@ -64,6 +65,8 @@ class Stats:
     running_lora_adapters: List[str]
     max_lora: str
 
+    spec_decode_metrics: Optional["SpecDecodeWorkerMetrics"] = None
+
 
 class StatLoggerBase(ABC):
     """Base class for StatLogger."""
@@ -74,6 +77,7 @@ class StatLoggerBase(ABC):
         self.num_generation_tokens: List[int] = []
         self.last_local_log = time.time()
         self.local_interval = local_interval
+        self.spec_decode_metrics: Optional[SpecDecodeWorkerMetrics] = None
 
     @abstractmethod
     def log(self, stats: Stats) -> None:
@@ -82,3 +86,9 @@ class StatLoggerBase(ABC):
     @abstractmethod
     def info(self, type: str, obj: SupportsMetricsInfo) -> None:
         raise NotImplementedError
+
+    def maybe_update_spec_decode_metrics(self, stats: Stats):
+        """Save spec decode metrics (since they are unlikely
+        to be emitted at same time as log interval)."""
+        if stats.spec_decode_metrics is not None:
+            self.spec_decode_metrics = stats.spec_decode_metrics

