diff --git a/vllm/v1/worker/lora_model_runner_mixin.py b/vllm/v1/worker/lora_model_runner_mixin.py
index e416f5032..afa41a37e 100644
--- a/vllm/v1/worker/lora_model_runner_mixin.py
+++ b/vllm/v1/worker/lora_model_runner_mixin.py
@@ -5,23 +5,17 @@ Define LoRA functionality mixin for model runners.
 """
 
 from contextlib import contextmanager
-from typing import Optional, Union
 
 import numpy as np
-import torch
 import torch.nn as nn
 
-from vllm.config import VllmConfig
-from vllm.config.lora import LoRAConfig
+from vllm.config import LoRAConfig, ModelConfig, SchedulerConfig
 from vllm.logger import init_logger
 from vllm.lora.layers import LoRAMapping
 from vllm.lora.request import LoRARequest
 from vllm.lora.worker_manager import LRUCacheWorkerLoRAManager
 from vllm.model_executor.models import supports_lora, supports_multimodal
-from vllm.v1.worker.gpu_input_batch import InputBatch as GPUInputBatch
-from vllm.v1.worker.tpu_input_batch import InputBatch as TPUInputBatch
-
-InputBatch = Union[TPUInputBatch, GPUInputBatch]
+from vllm.v1.worker.gpu_input_batch import InputBatch
 
 logger = init_logger(__name__)
 
@@ -31,8 +25,9 @@ class LoRAModelRunnerMixin:
 
     LORA_WARMUP_RANK = 8
 
-    def load_lora_model(self, model: nn.Module, vllm_config: VllmConfig,
-                        device: torch.device) -> nn.Module:
+    def load_lora_model(self, model: nn.Module, model_config: ModelConfig,
+                        scheduler_config: SchedulerConfig,
+                        lora_config: LoRAConfig, device: str) -> nn.Module:
 
         if not supports_lora(model):
             raise ValueError(
@@ -42,19 +37,27 @@ class LoRAModelRunnerMixin:
             logger.warning("Regarding multimodal models, vLLM currently "
                            "only supports adding LoRA to language model.")
 
+        # Use get_text_config() in case of multimodal models
+        text_config = model_config.hf_config.get_text_config()
+
         # Add LoRA Manager to the Model Runner
         self.lora_manager = LRUCacheWorkerLoRAManager(
-            vllm_config,
+            scheduler_config.max_num_seqs,
+            scheduler_config.max_num_batched_tokens,
+            model_config.get_vocab_size(),
+            lora_config,
             device,
             model.embedding_modules,
             model.embedding_padding_modules,
+            max_position_embeddings=text_config.max_position_embeddings,
         )
         return self.lora_manager.create_lora_manager(model)
 
     def _set_active_loras(self, prompt_lora_mapping: tuple[int, ...],
                           token_lora_mapping: tuple[int, ...],
                           lora_requests: set[LoRARequest]) -> None:
-        self._ensure_lora_enabled()
+        if not self.lora_manager:
+            raise RuntimeError("LoRA is not enabled.")
 
         # Set is_prefill to True, so we always use the SGMV kernels on
         # non-cuda platforms.
@@ -65,11 +68,6 @@ class LoRAModelRunnerMixin:
                                    is_prefill=True)
         self.lora_manager.set_active_adapters(lora_requests, lora_mapping)
 
-    def _ensure_lora_enabled(self) -> None:
-        if not hasattr(self, "lora_manager"):
-            raise RuntimeError(
-                "LoRA is not enabled. Use --enable-lora to enable LoRA.")
-
     def set_active_loras(self, input_batch: InputBatch,
                          num_scheduled_tokens: np.ndarray) -> None:
 
@@ -83,9 +81,7 @@ class LoRAModelRunnerMixin:
                                       lora_requests)
 
     @contextmanager
-    def maybe_setup_dummy_loras(self,
-                                lora_config: Optional[LoRAConfig],
-                                remove_lora: bool = True):
+    def maybe_setup_dummy_loras(self, lora_config):
         if lora_config is None:
             yield
         else:
@@ -112,11 +108,10 @@ class LoRAModelRunnerMixin:
                 yield
 
             # __exit__ code
-            if remove_lora:
-                self.lora_manager.remove_all_adapters()
+            self.lora_manager.remove_all_adapters()
 
     @contextmanager
-    def maybe_select_dummy_loras(self, lora_config: Optional[LoRAConfig],
+    def maybe_select_dummy_loras(self, lora_config: LoRAConfig,
                                  num_scheduled_tokens: np.ndarray):
         if lora_config is None:
             yield
@@ -150,34 +145,29 @@ class LoRAModelRunnerMixin:
             yield
 
     @contextmanager
-    def maybe_dummy_run_with_lora(self,
-                                  lora_config: Optional[LoRAConfig],
-                                  num_scheduled_tokens: np.ndarray,
-                                  remove_lora: bool = True):
-        with (
-                self.maybe_setup_dummy_loras(lora_config, remove_lora),
-                self.maybe_select_dummy_loras(lora_config,
-                                              num_scheduled_tokens),
-        ):
+    def maybe_dummy_run_with_lora(self, lora_config: LoRAConfig,
+                                  num_scheduled_tokens: np.ndarray):
+        with self.maybe_setup_dummy_loras(
+                lora_config), self.maybe_select_dummy_loras(
+                    lora_config, num_scheduled_tokens):
             yield
 
-    def maybe_remove_all_loras(self, lora_config: Optional[LoRAConfig]):
-        if lora_config is None:
-            return
-        self.lora_manager.remove_all_adapters()
-
     def add_lora(self, lora_request: LoRARequest) -> bool:
-        self._ensure_lora_enabled()
+        if not self.lora_manager:
+            raise RuntimeError("LoRA is not enabled.")
         return self.lora_manager.add_adapter(lora_request)
 
     def remove_lora(self, lora_id: int) -> bool:
-        self._ensure_lora_enabled()
+        if not self.lora_manager:
+            raise RuntimeError("LoRA is not enabled.")
         return self.lora_manager.remove_adapter(lora_id)
 
     def pin_lora(self, lora_id: int) -> bool:
-        self._ensure_lora_enabled()
+        if not self.lora_manager:
+            raise RuntimeError("LoRA is not enabled.")
         return self.lora_manager.pin_adapter(lora_id)
 
     def list_loras(self) -> set[int]:
-        self._ensure_lora_enabled()
+        if not self.lora_manager:
+            raise RuntimeError("LoRA is not enabled.")
         return self.lora_manager.list_adapters()

