diff --git a/vllm/distributed/kv_transfer/kv_transfer_state.py b/vllm/distributed/kv_transfer/kv_transfer_state.py
index d5747bed9..60f1d5d8b 100644
--- a/vllm/distributed/kv_transfer/kv_transfer_state.py
+++ b/vllm/distributed/kv_transfer/kv_transfer_state.py
@@ -8,6 +8,7 @@ from vllm.distributed.kv_transfer.kv_connector.factory import (
     KVConnectorFactory)
 from vllm.distributed.kv_transfer.kv_connector.v1 import (KVConnectorBase_V1,
                                                           KVConnectorRole)
+from vllm.distributed.parallel_state import get_world_group
 
 if TYPE_CHECKING:
     from vllm.config import VllmConfig
@@ -60,14 +61,11 @@ def ensure_kv_transfer_initialized(vllm_config: "VllmConfig") -> None:
     if (vllm_config.kv_transfer_config.is_kv_transfer_instance
             and _KV_CONNECTOR_AGENT is None):
         if envs.VLLM_USE_V1:
-            _KV_CONNECTOR_AGENT = KVConnectorFactory.create_connector(
+            _KV_CONNECTOR_AGENT = KVConnectorFactory.create_connector_v1(
                 config=vllm_config, role=KVConnectorRole.WORKER)
         else:
-            raise ValueError("V0 is no longer supported")
-
-
-def ensure_kv_transfer_shutdown() -> None:
-    global _KV_CONNECTOR_AGENT
-    if _KV_CONNECTOR_AGENT is not None:
-        _KV_CONNECTOR_AGENT.shutdown()
-        _KV_CONNECTOR_AGENT = None
+            _KV_CONNECTOR_AGENT = KVConnectorFactory.create_connector_v0(
+                rank=get_world_group().rank,
+                local_rank=get_world_group().local_rank,
+                config=vllm_config,
+            )

