diff --git a/megatron/core/timers.py b/megatron/core/timers.py
index a34b6634..cf31e6a7 100644
--- a/megatron/core/timers.py
+++ b/megatron/core/timers.py
@@ -136,7 +136,7 @@ class Timer(TimerBase):
         assert not self._started, 'timer has already been started'
         if barrier:
             torch.distributed.barrier(group=self._barrier_group)
-        torch.cuda.synchronize()
+        torch.musa.synchronize()
         self._start_time = time.time()
         self._started = True
 
@@ -149,7 +149,7 @@ class Timer(TimerBase):
         assert self._started, 'timer is not started'
         if barrier:
             torch.distributed.barrier(group=self._barrier_group)
-        torch.cuda.synchronize()
+        torch.musa.synchronize()
         elapsed = time.time() - self._start_time
         self._elapsed += elapsed
         self._active_time += elapsed
@@ -273,7 +273,7 @@ class Timers:
         # and since we are only gathering a small amount of data,
         # it should be ok to use all-gather instead of gather.
         rank_name_to_time = torch.zeros(
-            (world_size, len(names)), dtype=torch.float, device=torch.cuda.current_device()
+            (world_size, len(names)), dtype=torch.float, device=torch.musa.current_device()
         )
         for i, name in enumerate(names):
             if name in self._timers:

