diff --git a/vllm/model_executor/layers/quantization/utils/bitblas_utils.py b/vllm/model_executor/layers/quantization/utils/bitblas_utils.py
index 4c2e54873..82ee3edfd 100644
--- a/vllm/model_executor/layers/quantization/utils/bitblas_utils.py
+++ b/vllm/model_executor/layers/quantization/utils/bitblas_utils.py
@@ -3,7 +3,6 @@
 from typing import Optional
 
 import torch
-from packaging import version
 
 from vllm.platforms import current_platform
 from vllm.scalar_type import ScalarType, scalar_types
@@ -76,8 +75,7 @@ def _check_bitblas_supported(
     # Finally, check if bitblas is installed
     try:
         import bitblas
-        if version.parse(
-                bitblas.__version__) < version.parse(MINIMUM_BITBLAS_VERSION):
+        if bitblas.__version__ < MINIMUM_BITBLAS_VERSION:
             raise ImportError("bitblas version is wrong. Please "
                               f"install bitblas>={MINIMUM_BITBLAS_VERSION}")
     except ImportError:

