diff --git a/vllm/model_executor/layers/quantization/utils/marlin_utils_test.py b/vllm/model_executor/layers/quantization/utils/marlin_utils_test.py
index f5acd03cc..b2c228c24 100644
--- a/vllm/model_executor/layers/quantization/utils/marlin_utils_test.py
+++ b/vllm/model_executor/layers/quantization/utils/marlin_utils_test.py
@@ -19,7 +19,7 @@ class MarlinWorkspace:
 
     def __init__(self, out_features, min_thread_n, max_parallel):
         assert (out_features % min_thread_n == 0), (
-            "out_features = {} is indivisible by min_thread_n = {}".format(
+            "out_features = {} is undivisible by min_thread_n = {}".format(
                 out_features, min_thread_n))
 
         max_workspace_size = ((out_features // min_thread_n) * max_parallel)

