diff --git a/vllm/model_executor/layers/quantization/utils/w8a8_utils.py b/vllm/model_executor/layers/quantization/utils/w8a8_utils.py
index b434b7acf..adc67aa64 100644
--- a/vllm/model_executor/layers/quantization/utils/w8a8_utils.py
+++ b/vllm/model_executor/layers/quantization/utils/w8a8_utils.py
@@ -4,17 +4,11 @@
 from typing import Callable, Optional, Union
 
 import torch
-from packaging import version
 
 from vllm import _custom_ops as ops
 from vllm import envs
 from vllm.config import CompilationLevel, get_current_vllm_config
-from vllm.model_executor.layers.quantization.input_quant_fp8 import QuantFP8
-from vllm.model_executor.layers.quantization.utils.quant_utils import (
-    GroupShape)
 from vllm.platforms import current_platform
-from vllm.utils import direct_register_custom_op
-from vllm.utils.flashinfer import flashinfer_scaled_fp8_mm, has_flashinfer
 
 # Input scaling factors are no longer optional in _scaled_mm starting
 # from pytorch 2.5. Allocating a dummy tensor to pass as input_scale
@@ -23,9 +17,9 @@ TORCH_DEVICE_IDENTITY = None
 # The condition to determine if it is on a platform that supports
 # torch._scaled_mm rowwise feature.
 # The condition is determined once as the operations
-# are time-consuming.
-USE_ROWWISE_TORCH_SCALED_MM = (current_platform.is_rocm() and version.parse(
-    torch.__version__) >= version.parse("2.7")
+# are time consuming.
+USE_ROWWISE_TORCH_SCALED_MM = (current_platform.is_rocm()
+                               and torch.__version__[0:3] >= "2.7"
                                and current_platform.has_device_capability(94))
 
 
@@ -123,9 +117,6 @@ def requantize_with_max_scale(
     if unfused_module_in_checkpoint:
         start = 0
         for idx, logical_width in enumerate(logical_widths):
-            # Skip any component with zero width.
-            if logical_width == 0:
-                continue
             end = start + logical_width
             weight_dq = per_tensor_dequantize(weight[start:end, :],
                                               weight_scale[idx])
@@ -158,32 +149,18 @@ def cutlass_w8a8_scaled_mm(*, qinput: torch.Tensor, weight: torch.Tensor,
     return output.view(*output_shape)
 
 
-def flashinfer_w8a8_scaled_mm(*, qinput: torch.Tensor, weight: torch.Tensor,
-                              out_dtype: torch.dtype, scale_a: torch.Tensor,
-                              scale_b: torch.Tensor, bias: torch.Tensor,
-                              output_shape: list, **kwargs) -> torch.Tensor:
-
-    return flashinfer_scaled_fp8_mm(qinput,
-                                    weight,
-                                    out_dtype=out_dtype,
-                                    scale_a=scale_a,
-                                    scale_b=scale_b,
-                                    bias=bias)
-
-
-def rocm_per_tensor_w8a8_scaled_mm_impl(qinput: torch.Tensor,
-                                        weight: torch.Tensor,
-                                        out_dtype: torch.dtype,
-                                        scale_a: torch.Tensor,
-                                        scale_b: torch.Tensor,
-                                        bias: torch.Tensor) -> torch.Tensor:
+def rocm_per_tensor_w8a8_scaled_mm(*, qinput: torch.Tensor,
+                                   weight: torch.Tensor,
+                                   out_dtype: torch.dtype,
+                                   scale_a: torch.Tensor,
+                                   scale_b: torch.Tensor, bias: torch.Tensor,
+                                   input_2d: torch.Tensor,
+                                   output_shape: list) -> torch.Tensor:
     from vllm.platforms.rocm import on_mi3xx
-    if envs.VLLM_ROCM_USE_SKINNY_GEMM and on_mi3xx() and \
-            qinput.shape[0] == 1 and \
-            qinput.shape[1] % 16 == 0 and \
-            ((bias is None) or (bias.dtype == out_dtype)) :
+    if envs.VLLM_ROCM_USE_SKINNY_GEMM and on_mi3xx(
+    ) and qinput.shape[0] == 1 and qinput.shape[1] % 16 == 0:
         output = ops.wvSplitKQ(weight.t(), qinput, out_dtype, scale_a, scale_b,
-                               current_platform.get_cu_count(), bias)
+                               current_platform.get_cu_count())
     else:
         output = torch._scaled_mm(qinput,
                                   weight,
@@ -191,35 +168,8 @@ def rocm_per_tensor_w8a8_scaled_mm_impl(qinput: torch.Tensor,
                                   scale_a=scale_a,
                                   scale_b=scale_b,
                                   bias=bias)
-    return output
-
-
-def rocm_per_tensor_w8a8_scaled_mm_fake(qinput: torch.Tensor,
-                                        weight: torch.Tensor,
-                                        out_dtype: torch.dtype,
-                                        scale_a: torch.Tensor,
-                                        scale_b: torch.Tensor,
-                                        bias: torch.Tensor) -> torch.Tensor:
-    return qinput.new_empty((*qinput.shape[:-1], weight.shape[1]),
-                            dtype=out_dtype)
 
-
-def rocm_per_tensor_w8a8_scaled_mm(*, qinput: torch.Tensor,
-                                   weight: torch.Tensor,
-                                   out_dtype: torch.dtype,
-                                   scale_a: torch.Tensor,
-                                   scale_b: torch.Tensor, bias: torch.Tensor,
-                                   output_shape: list) -> torch.Tensor:
-    output = torch.ops.vllm.rocm_per_tensor_w8a8_scaled_mm_impl(
-        qinput, weight, out_dtype, scale_a, scale_b, bias)
-    return torch.narrow(output, 0, 0, qinput.shape[0]).view(*output_shape)
-
-
-direct_register_custom_op(
-    op_name="rocm_per_tensor_w8a8_scaled_mm_impl",
-    op_func=rocm_per_tensor_w8a8_scaled_mm_impl,
-    fake_impl=rocm_per_tensor_w8a8_scaled_mm_fake,
-)
+    return torch.narrow(output, 0, 0, input_2d.shape[0]).view(*output_shape)
 
 
 def torch_per_tensor_w8a8_scaled_mm(*, qinput: torch.Tensor,
@@ -227,6 +177,7 @@ def torch_per_tensor_w8a8_scaled_mm(*, qinput: torch.Tensor,
                                     out_dtype: torch.dtype,
                                     scale_a: torch.Tensor,
                                     scale_b: torch.Tensor, bias: torch.Tensor,
+                                    input_2d: torch.Tensor,
                                     output_shape: list) -> torch.Tensor:
     output = torch._scaled_mm(qinput,
                               weight,
@@ -239,7 +190,7 @@ def torch_per_tensor_w8a8_scaled_mm(*, qinput: torch.Tensor,
     if type(output) is tuple and len(output) == 2:
         output = output[0]
 
-    return torch.narrow(output, 0, 0, qinput.shape[0]).view(*output_shape)
+    return torch.narrow(output, 0, 0, input_2d.shape[0]).view(*output_shape)
 
 
 def torch_per_token_w8a8_scaled_mm(*, qinput: torch.Tensor,
@@ -247,8 +198,8 @@ def torch_per_token_w8a8_scaled_mm(*, qinput: torch.Tensor,
                                    out_dtype: torch.dtype,
                                    scale_a: torch.Tensor,
                                    scale_b: torch.Tensor, bias: torch.Tensor,
-                                   output_shape: list,
-                                   **kwargs) -> torch.Tensor:
+                                   input_2d: torch.Tensor,
+                                   output_shape: list) -> torch.Tensor:
     # Note: Callers of this function should check USE_ROWWISE_TORCH_SCALED_MM
     #  when using it.
     #  For now it has only been validated on ROCm platform.
@@ -267,7 +218,7 @@ def torch_per_token_w8a8_scaled_mm(*, qinput: torch.Tensor,
                               scale_b=scale_b.t(),
                               bias=bias)
 
-    output = torch.narrow(output, 0, 0, qinput.shape[0])
+    output = torch.narrow(output, 0, 0, input_2d.shape[0])
     output = output.view(*output_shape)
     return output
 
@@ -277,6 +228,7 @@ def torch_channelwise_w8a8_scaled_mm(*, qinput: torch.Tensor,
                                      out_dtype: torch.dtype,
                                      scale_a: torch.Tensor,
                                      scale_b: torch.Tensor, bias: torch.Tensor,
+                                     input_2d: torch.Tensor,
                                      output_shape: list,
                                      **kwargs) -> torch.Tensor:
     # Use unfused DQ due to limitations with scaled_mm
@@ -306,8 +258,8 @@ def torch_channelwise_w8a8_scaled_mm(*, qinput: torch.Tensor,
     if type(output) is tuple and len(output) == 2:
         output = output[0]
     # Unpad (undo num_token_padding)
-    output = torch.narrow(output, 0, 0, qinput.shape[0])
-    x_scale = torch.narrow(scale_a, 0, 0, qinput.shape[0])
+    output = torch.narrow(output, 0, 0, input_2d.shape[0])
+    x_scale = torch.narrow(scale_a, 0, 0, input_2d.shape[0])
 
     # DQ
     # C = sw * sx * (X * W) + bias
@@ -318,28 +270,21 @@ def torch_channelwise_w8a8_scaled_mm(*, qinput: torch.Tensor,
 
 
 def dispatch_w8a8_scaled_mm(
-        preferred_backend: str, per_tensor_weights: bool,
-        per_tensor_activations: bool) -> Callable[..., torch.Tensor]:
+        cutlass_fp8_supported: bool, per_tensor_weights: bool,
+        per_tensor_activations: bool, use_per_token_if_dynamic: Optional[bool]
+) -> Callable[..., torch.Tensor]:
 
+    if cutlass_fp8_supported:
+        return cutlass_w8a8_scaled_mm
     if per_tensor_weights and per_tensor_activations:
-        if preferred_backend == "rocm":
+        if current_platform.is_rocm():
             return rocm_per_tensor_w8a8_scaled_mm
-        if preferred_backend == "flashinfer":
-            return flashinfer_w8a8_scaled_mm
-        if preferred_backend == "cutlass":
-            return cutlass_w8a8_scaled_mm
         return torch_per_tensor_w8a8_scaled_mm
-
-    # cutlass_scaled_mm supports per tensor/channel W and per tensor/token A
-    if preferred_backend == "cutlass" or preferred_backend == "flashinfer":
-        return cutlass_w8a8_scaled_mm
-
-    # If torch.scaled_mm supports per-channel (weights) per-token (inputs)
-    if not per_tensor_weights and not per_tensor_activations \
-            and USE_ROWWISE_TORCH_SCALED_MM:
-        return torch_per_token_w8a8_scaled_mm
-    # Normally, torch.scaled_mm supports per tensor weights + activations only
+    # torch.scaled_mm supports per tensor weights + activations only
     # so fallback to naive if per channel or per token
+    if (use_per_token_if_dynamic and not per_tensor_weights
+            and not per_tensor_activations and USE_ROWWISE_TORCH_SCALED_MM):
+        return torch_per_token_w8a8_scaled_mm
     return torch_channelwise_w8a8_scaled_mm
 
 
@@ -354,19 +299,11 @@ class Fp8LinearOp:
     """
 
     def __init__(self,
-                 act_quant_static: bool,
-                 act_quant_group_shape: GroupShape = GroupShape.PER_TENSOR,
+                 cutlass_fp8_supported: bool = cutlass_fp8_supported(),
+                 use_per_token_if_dynamic: bool = False,
                  pad_output: Optional[bool] = None):
-        if current_platform.is_rocm():
-            self.preferred_backend = "rocm"
-        elif current_platform.is_cuda() and cutlass_fp8_supported():
-            if has_flashinfer() and current_platform.has_device_capability(
-                    100):
-                self.preferred_backend = "flashinfer"
-            else:
-                self.preferred_backend = "cutlass"
-        else:
-            self.preferred_backend = "torch"
+        self.cutlass_fp8_supported = cutlass_fp8_supported
+        self.use_per_token_if_dynamic = use_per_token_if_dynamic
 
         # Note: we pad the input because torch._scaled_mm is more performant
         # for matrices with batch dimension > 16.
@@ -375,15 +312,9 @@ class Fp8LinearOp:
         # as it breaks with dynamic shapes.
         if pad_output is None:
             config = get_current_vllm_config().compilation_config
-            pad_output = config.level < CompilationLevel.PIECEWISE and \
-                         self.preferred_backend == "torch"
-
-        self.output_padding = 17 if pad_output else None
-        self.act_quant_static = act_quant_static
-        self.act_quant_group_shape = act_quant_group_shape
-        self.quant_fp8 = QuantFP8(static=act_quant_static,
-                                  group_shape=act_quant_group_shape,
-                                  num_token_padding=self.output_padding)
+            pad_output = config.level < CompilationLevel.PIECEWISE
+        self.output_padding = 17 if (
+            pad_output and not current_platform.is_rocm()) else None
 
     def apply(
         self,
@@ -394,6 +325,8 @@ class Fp8LinearOp:
         input_scale: Optional[torch.Tensor] = None,
         input_scale_ub: Optional[torch.Tensor] = None,
         bias: Optional[torch.Tensor] = None,
+        # TODO(luka) remove this parameter in favor of __init__
+        use_per_token_if_dynamic: Optional[bool] = None
     ) -> torch.Tensor:
         # ops.scaled_fp8_quant supports both dynamic and static quant.
         #   If dynamic, layer.input_scale is None and x_scale computed from x.
@@ -403,27 +336,40 @@ class Fp8LinearOp:
         input_2d = input.view(-1, input.shape[-1])
         output_shape = [*input.shape[:-1], weight.shape[1]]
 
+        # TODO(luka) this is here because currently MLA only decides this
+        #  during the forward method instead of in __init__.
+        if use_per_token_if_dynamic is None:
+            use_per_token_if_dynamic = self.use_per_token_if_dynamic
+
         if out_dtype is None:
             out_dtype = input.dtype
 
-        # If input not quantized
-        # TODO(luka) remove this path if not used anymore
-        if input.dtype != current_platform.fp8_dtype():
-            qinput, x_scale = self.quant_fp8(
+        # cutlass_scaled_mm supports per tensor/channel W and per tensor/token A
+        if self.cutlass_fp8_supported:
+            assert input.dtype != current_platform.fp8_dtype(
+            ), "FP8 input to cutlass is not currently implemented"
+            qinput, x_scale = ops.scaled_fp8_quant(
                 input_2d,
                 input_scale,
-                input_scale_ub,
-            )
+                scale_ub=input_scale_ub,
+                use_per_token_if_dynamic=use_per_token_if_dynamic)
         else:
-            qinput, x_scale = input_2d, input_scale
+            if input.dtype != current_platform.fp8_dtype():
+                # Maybe apply padding to output, see comment in __init__
+                qinput, x_scale = ops.scaled_fp8_quant(
+                    input_2d,
+                    input_scale,
+                    num_token_padding=self.output_padding,
+                    use_per_token_if_dynamic=use_per_token_if_dynamic)
+            else:
+                qinput, x_scale = input_2d, input_scale
 
         per_tensor_weights = (weight_scale.numel() == 1)
         per_tensor_activations = (x_scale.numel() == 1)
 
-        # TODO(luka) do this dispatch during init (after ScaledMM refactor)
-        w8a8_scaled_mm_func = dispatch_w8a8_scaled_mm(self.preferred_backend,
-                                                      per_tensor_weights,
-                                                      per_tensor_activations)
+        w8a8_scaled_mm_func = dispatch_w8a8_scaled_mm(
+            self.cutlass_fp8_supported, per_tensor_weights,
+            per_tensor_activations, use_per_token_if_dynamic)
 
         return w8a8_scaled_mm_func(qinput=qinput,
                                    weight=weight,
@@ -431,6 +377,7 @@ class Fp8LinearOp:
                                    scale_a=x_scale,
                                    scale_b=weight_scale,
                                    bias=bias,
+                                   input_2d=input_2d,
                                    output_shape=output_shape)
 
 

